<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Matthew Stephens" />

<meta name="date" content="2025-03-12" />

<title>eb_snmu</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephens999/misc">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">eb_snmu</h1>
<h4 class="author">Matthew Stephens</h4>
<h4 class="date">2025-03-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree87489896d3d4c467d38550b4c76928cc83708d0btargetblank8748989a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephens999/misc/tree/87489896d3d4c467d38550b4c76928cc83708d0b" target="_blank">8748989</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree87489896d3d4c467d38550b4c76928cc83708d0btargetblank8748989a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephens999/misc/tree/87489896d3d4c467d38550b4c76928cc83708d0b" target="_blank">8748989</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/ALStruct_cache/
    Ignored:    data/.Rhistory
    Ignored:    data/methylation-data-for-matthew.rds
    Ignored:    data/pbmc/
    Ignored:    data/pbmc_purified.RData

Untracked files:
    Untracked:  .dropbox
    Untracked:  Icon
    Untracked:  analysis/GHstan.Rmd
    Untracked:  analysis/GTEX-cogaps.Rmd
    Untracked:  analysis/PACS.Rmd
    Untracked:  analysis/Rplot.png
    Untracked:  analysis/SPCAvRP.rmd
    Untracked:  analysis/abf_comparisons.Rmd
    Untracked:  analysis/admm_02.Rmd
    Untracked:  analysis/admm_03.Rmd
    Untracked:  analysis/bispca.Rmd
    Untracked:  analysis/cache/
    Untracked:  analysis/cholesky.Rmd
    Untracked:  analysis/compare-transformed-models.Rmd
    Untracked:  analysis/cormotif.Rmd
    Untracked:  analysis/cp_ash.Rmd
    Untracked:  analysis/eQTL.perm.rand.pdf
    Untracked:  analysis/eb_prepilot.Rmd
    Untracked:  analysis/eb_var.Rmd
    Untracked:  analysis/ebpmf1.Rmd
    Untracked:  analysis/ebpmf_sla_text.Rmd
    Untracked:  analysis/ebspca_sims.Rmd
    Untracked:  analysis/explore_psvd.Rmd
    Untracked:  analysis/fa_check_identify.Rmd
    Untracked:  analysis/fa_iterative.Rmd
    Untracked:  analysis/flash_cov_overlapping_groups_init.Rmd
    Untracked:  analysis/flash_test_tree.Rmd
    Untracked:  analysis/flashier_newgroups.Rmd
    Untracked:  analysis/flashier_nmf_triples.Rmd
    Untracked:  analysis/flashier_pbmc.Rmd
    Untracked:  analysis/flashier_snn_shifted_prior.Rmd
    Untracked:  analysis/greedy_ebpmf_exploration_00.Rmd
    Untracked:  analysis/ieQTL.perm.rand.pdf
    Untracked:  analysis/lasso_em_03.Rmd
    Untracked:  analysis/m6amash.Rmd
    Untracked:  analysis/mash_bhat_z.Rmd
    Untracked:  analysis/mash_ieqtl_permutations.Rmd
    Untracked:  analysis/methylation_example.Rmd
    Untracked:  analysis/mixsqp.Rmd
    Untracked:  analysis/mr.ash_lasso_init.Rmd
    Untracked:  analysis/mr.mash.test.Rmd
    Untracked:  analysis/mr_ash_modular.Rmd
    Untracked:  analysis/mr_ash_parameterization.Rmd
    Untracked:  analysis/mr_ash_ridge.Rmd
    Untracked:  analysis/mv_gaussian_message_passing.Rmd
    Untracked:  analysis/nejm.Rmd
    Untracked:  analysis/nmf_bg.Rmd
    Untracked:  analysis/nonneg_underapprox.Rmd
    Untracked:  analysis/normal_conditional_on_r2.Rmd
    Untracked:  analysis/normalize.Rmd
    Untracked:  analysis/pbmc.Rmd
    Untracked:  analysis/pca_binary_weighted.Rmd
    Untracked:  analysis/pca_l1.Rmd
    Untracked:  analysis/poisson_nmf_approx.Rmd
    Untracked:  analysis/poisson_shrink.Rmd
    Untracked:  analysis/poisson_transform.Rmd
    Untracked:  analysis/qrnotes.txt
    Untracked:  analysis/ridge_iterative_02.Rmd
    Untracked:  analysis/ridge_iterative_splitting.Rmd
    Untracked:  analysis/samps/
    Untracked:  analysis/sc_bimodal.Rmd
    Untracked:  analysis/shrinkage_comparisons_changepoints.Rmd
    Untracked:  analysis/susie_cov.Rmd
    Untracked:  analysis/susie_en.Rmd
    Untracked:  analysis/susie_z_investigate.Rmd
    Untracked:  analysis/svd-timing.Rmd
    Untracked:  analysis/temp.RDS
    Untracked:  analysis/temp.Rmd
    Untracked:  analysis/test-figure/
    Untracked:  analysis/test.Rmd
    Untracked:  analysis/test.Rpres
    Untracked:  analysis/test.md
    Untracked:  analysis/test_qr.R
    Untracked:  analysis/test_sparse.Rmd
    Untracked:  analysis/tree_dist_top_eigenvector.Rmd
    Untracked:  analysis/z.txt
    Untracked:  code/multivariate_testfuncs.R
    Untracked:  code/rqb.hacked.R
    Untracked:  data/4matthew/
    Untracked:  data/4matthew2/
    Untracked:  data/E-MTAB-2805.processed.1/
    Untracked:  data/ENSG00000156738.Sim_Y2.RDS
    Untracked:  data/GDS5363_full.soft.gz
    Untracked:  data/GSE41265_allGenesTPM.txt
    Untracked:  data/Muscle_Skeletal.ACTN3.pm1Mb.RDS
    Untracked:  data/P.rds
    Untracked:  data/Thyroid.FMO2.pm1Mb.RDS
    Untracked:  data/bmass.HaemgenRBC2016.MAF01.Vs2.MergedDataSources.200kRanSubset.ChrBPMAFMarkerZScores.vs1.txt.gz
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.NewSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.PreviousSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/eb_prepilot/
    Untracked:  data/finemap_data/fmo2.sim/b.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/data
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.ld
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.z
    Untracked:  data/finemap_data/fmo2.sim/pos.txt
    Untracked:  data/logm.csv
    Untracked:  data/m.cd.RDS
    Untracked:  data/m.cdu.old.RDS
    Untracked:  data/m.new.cd.RDS
    Untracked:  data/m.old.cd.RDS
    Untracked:  data/mainbib.bib.old
    Untracked:  data/mat.csv
    Untracked:  data/mat.txt
    Untracked:  data/mat_new.csv
    Untracked:  data/matrix_lik.rds
    Untracked:  data/paintor_data/
    Untracked:  data/running_data_chris.csv
    Untracked:  data/running_data_matthew.csv
    Untracked:  data/temp.txt
    Untracked:  data/y.txt
    Untracked:  data/y_f.txt
    Untracked:  data/zscore_jointLCLs_m6AQTLs_susie_eQTLpruned.rds
    Untracked:  data/zscore_jointLCLs_random.rds
    Untracked:  explore_udi.R
    Untracked:  output/fit.k10.rds
    Untracked:  output/fit.nn.pbmc.purified.rds
    Untracked:  output/fit.nn.rds
    Untracked:  output/fit.nn.s.001.rds
    Untracked:  output/fit.nn.s.01.rds
    Untracked:  output/fit.nn.s.1.rds
    Untracked:  output/fit.nn.s.10.rds
    Untracked:  output/fit.snn.s.001.rds
    Untracked:  output/fit.snn.s.01.nninit.rds
    Untracked:  output/fit.snn.s.01.rds
    Untracked:  output/fit.varbvs.RDS
    Untracked:  output/fit2.nn.pbmc.purified.rds
    Untracked:  output/glmnet.fit.RDS
    Untracked:  output/snn07.txt
    Untracked:  output/snn34.txt
    Untracked:  output/test.bv.txt
    Untracked:  output/test.gamma.txt
    Untracked:  output/test.hyp.txt
    Untracked:  output/test.log.txt
    Untracked:  output/test.param.txt
    Untracked:  output/test2.bv.txt
    Untracked:  output/test2.gamma.txt
    Untracked:  output/test2.hyp.txt
    Untracked:  output/test2.log.txt
    Untracked:  output/test2.param.txt
    Untracked:  output/test3.bv.txt
    Untracked:  output/test3.gamma.txt
    Untracked:  output/test3.hyp.txt
    Untracked:  output/test3.log.txt
    Untracked:  output/test3.param.txt
    Untracked:  output/test4.bv.txt
    Untracked:  output/test4.gamma.txt
    Untracked:  output/test4.hyp.txt
    Untracked:  output/test4.log.txt
    Untracked:  output/test4.param.txt
    Untracked:  output/test5.bv.txt
    Untracked:  output/test5.gamma.txt
    Untracked:  output/test5.hyp.txt
    Untracked:  output/test5.log.txt
    Untracked:  output/test5.param.txt

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/ebnm_binormal.Rmd
    Modified:   analysis/flashier_log1p.Rmd
    Modified:   analysis/flashier_sla_text.Rmd
    Modified:   analysis/logistic_z_scores.Rmd
    Modified:   analysis/mr_ash_pen.Rmd
    Modified:   analysis/nmu_em.Rmd
    Modified:   analysis/susie_flash.Rmd
    Modified:   misc.Rproj

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/eb_snmu.Rmd</code>) and HTML
(<code>docs/eb_snmu.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephens999/misc/blob/87489896d3d4c467d38550b4c76928cc83708d0b/analysis/eb_snmu.Rmd" target="_blank">8748989</a>
</td>
<td>
Matthew Stephens
</td>
<td>
2025-03-21
</td>
<td>
workflowr::wflow_publish("eb_snmu.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(&quot;emg&quot;)</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: moments</code></pre>
<pre class="r"><code>library(&quot;ebnm&quot;)</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>I’m going to start putting together code for Empirical Bayes
symmetric non-negative matrix underapproximation (SNMU). Given an <span
class="math inline">\(n\times n\)</span> nonnegative matrix <span
class="math inline">\(S\)</span> the goal is to fit a model of the form
<span class="math display">\[S = vdv&#39; + b + e\]</span> with <span
class="math inline">\(v\)</span> an <span class="math inline">\(n \times
K\)</span> nonnegative matrix, <span class="math inline">\(d\)</span> is
a nonnegative diagonal matrix, and <span
class="math inline">\(b\)</span> is a nonnegative <span
class="math inline">\(n \times n\)</span> matrix. If <span
class="math inline">\(e\)</span> is small or 0 then <span
class="math inline">\(vdv&#39;\)</span> forms a nonnegative
underapproximation of <span class="math inline">\(S\)</span>. We assume
a prior <span class="math inline">\(g_k\)</span> on elements of <span
class="math inline">\(v[,k]\)</span>, which will be estimated by EB.</p>
<p>We assume elements of <span class="math inline">\(b\)</span> are iid
Exp(<span class="math inline">\(\lambda\)</span>) and elements of <span
class="math inline">\(e_{ij}\)</span> are <span
class="math inline">\(N(0,\sigma^2)\)</span>. Thus the elements of <span
class="math inline">\(S\)</span> have an Exponentially modified gaussian
(EMG) distribution.</p>
<p>This will use ideas from (ebpower.html) and (power_nneg.html), as
well as the ebnm function from (ebnm_binormal.html).</p>
<p>Note: while i started out with the underapproximation idea in mind,
there are reasons I started to focus on the simpler case b=0. These
include: convergence is slower with b&gt;0, and I am not sure I can
implement an efficient version for the b&gt;0 case when the original
matrix is sparse, which is a deal-breaker in practice. Of course this
code includes that case (b=0) as well.</p>
<pre class="r"><code># truncate and normalize a vector v
# returns a vector of 0s if all element of v are non-positive
trunc_and_norm = function(v){
  vplus = ifelse(v&gt;0,v,0)
  if(!all(vplus==0))
    vplus = vplus/sqrt(sum(vplus^2))
  return(vplus)
}

# returns the log likelihood of a matrix S under the
# Exponentially modified gaussian (EMG) model
loglik_emg = function(S,v,d,lambda,sigma){
  R = S - vdvt(v,d) 
  sum(demg(R,lambda=lambda,sigma=sigma,log=TRUE))
}</code></pre>
<div id="functions-for-ebnm_binormal" class="section level3">
<h3>Functions for ebnm_binormal</h3>
<pre class="r"><code>dbinormal = function (x,s,s0,lambda,log=TRUE){
  pi0 = 0.5
  pi1 = 0.5
  s2 = s^2
  s02 = s0^2
  l0 = dnorm(x,0,sqrt(lambda^2 * s02 + s2),log=TRUE)
  l1 = dnorm(x,lambda,sqrt(lambda^2 * s02 + s2),log=TRUE)
  logsum = log(pi0*exp(l0) + pi1*exp(l1))
 
  m = pmax(l0,l1)
  logsum = m + log(pi0*exp(l0-m) + pi1*exp(l1-m))
  if (log) return(sum(logsum))
  else return(exp(sum(logsum)))
}

ebnm_binormal = function(x,s){
  
  if(all(x&lt;0)){ # in this case optimal lambda is 0
    return(list(g = ashr::normalmix(pi=c(0.5,0.5), mean=c(0,lambda), sd=c(0,0)),posterior = data.frame(mean=rep(0,length(x)),sd=rep(0,length(x)))))
  }
  
  x = drop(x)
  s0 = 0.01
  lambda = optimize(function(lambda){-dbinormal(x,s,s0,lambda,log=TRUE)},
              lower = 0, upper = max(x))$minimum
  g = ashr::normalmix(pi=c(0.5,0.5), mean=c(0,lambda), sd=c(lambda * s0,lambda * s0))
  postmean = ashr::postmean(g,ashr::set_data(x,s))
  postsd = ashr::postsd(g,ashr::set_data(x,s))
  return(list(g = g, posterior = data.frame(mean=postmean,sd=postsd)))
}</code></pre>
</div>
<div id="functions-for-snmu" class="section level3">
<h3>Functions for SNMU</h3>
<pre class="r"><code># This is just the regular power update, used for initialization
power_update_r1 = function(S,v){
  newv = drop(S %*% v)
  if(!all(newv==0))
    v = newv/sqrt(sum(newv^2))
  return(v)
}

# initializes an SNMU fit for a matrix S
# initializes b = 0, v by random random normals, and d to be very small (1e-8). It runs n.power.iter iterations of the rank1 power update on each initialized v vector and then truncates and normalizes them.
# Note that the power update here is done with S, rather than S-VDV&#39;
snmu_init = function(S,K,lambda=1,sigma=0.1,b.init= matrix(0,nrow=nrow(S),ncol=ncol(S)), n.power.iter = 1){
  
  b = b.init
    # if b.init supplied, use it to estimate lambda and sigma
  # this maybe only makes sense if we use the low rank approx of S to estimate b
  if(!missing(b.init)){ 
    lambda = 1/mean(b)
    sigma = sqrt(mean((S-b)^2))
  }  
  
  v = matrix(rnorm(nrow(S)*K),ncol=K)
  for(k in 1:K){
    for(i in 1:n.power.iter){
      v[,k] = power_update_r1(S,v[,k])
    }
    v[,k] = trunc_and_norm(v[,k])
  }
    
  d = rep(1e-8,K)
  loglik = loglik_emg(S,v,d,lambda,sigma)
  
  return(list(S=S,b=b,v=v,d=d,lambda=lambda, sigma=sigma,loglik=loglik))
}

snmu_add_new_factor = function(fit, v = rnorm(nrow(fit$S)), n.power.iter = 1){
  for(i in 1:n.power.iter){
    v = power_update_r1(fit$S-fit$b-vdvt(fit$v,fit$d),v)
  }
  v = trunc_and_norm(v)
  
  fit$v = cbind(fit$v,v)
  fit$d = c(fit$d,1e-8)
  return(fit)
}

# return the last element of the vector loglik
snmu_loglik = function(fit){
  fit$loglik[length(fit$loglik)]
}

vdvt = function(v,d){
  v %*% diag(d,nrow=length(d)) %*% t(v)
}

snmu_update_d =function(fit,kset=1:ncol(fit$v)){
  fit = within(fit, {
    for(k in kset){
      U = v[,-k,drop=FALSE]
      D = diag(d[-k],nrow=length(d[-k]))
      d[k] = max(t(v[,k]) %*% (S - b) %*% v[,k] - t(v[,k]) %*% U %*% D %*% t(U) %*% v[,k],0)
    }
  })
}</code></pre>
<pre class="r"><code># note that this could be made more efficent for sparse S
# one puzzle is how to avoid evaluating b in cases where S =XX&#39; and X sparse...
# updates the columns of v in kset
snmu_update_v=function(fit,ebnm_fn,kset=1:ncol(fit$v)){
  fit = within(fit, {
    for(k in kset){
      
      U = v[,-k,drop=FALSE]
      D = diag(d[-k],nrow=length(d[-k]))
    
      newv = drop(S %*% v[,k,drop=FALSE] - b %*% v[,k,drop=FALSE] - U %*% D %*% t(U) %*% v[,k,drop=FALSE])
      
      if(!all(newv==0)){
        fit.ebnm = ebnm_fn(newv,sigma)        
        newv = fit.ebnm$posterior$mean
        if(!all(newv==0)){
          newv = newv/sqrt(sum(newv^2 + fit.ebnm$posterior$sd^2))
        } 
        rm(fit.ebnm)
      }
    
      v[,k] = newv
      d[k] = max(t(v[,k]) %*% (S - b) %*% v[,k] - t(v[,k]) %*% U %*% D %*% t(U) %*% v[,k],0)
    }
    rm(newv,D,U)
  })
  return(fit)
}


snmu_update_b=function(fit){
  fit = within(fit,{
    b = matrix(truncnorm::etruncnorm(a=0,mean= S-vdvt(v,d)-lambda*sigma^2,sd=sigma),nrow=nrow(S))
  })
  return(fit)
}

snmu_update_loglik=function(fit){
  fit$loglik = with(fit, c(loglik, loglik_emg(S,v,d,lambda,sigma)))
  return(fit)
}  

snmu_update_sigma_and_lambda_mom = function(fit){
  fit$lambda = with(fit, 1/mean(S-vdvt(v,d))) #method of moments estimates
  fit$sigma = with(fit, sd(S-vdvt(v,d)))
  return(fit)
}

snmu_update_sigma_mle = function(fit){
  fit.mle = mle(function(lsigma) {-loglik_emg(fit$S, fit$v, fit$d, fit$lambda,exp(lsigma))}, method = &quot;Brent&quot;, lower=c(-10), upper = c(10), start =c(log(fit$sigma)))
  fit$sigma = exp(coef(fit.mle)[1])
  return(fit)
}</code></pre>
</div>
</div>
<div id="experiments" class="section level2">
<h2>Experiments</h2>
<p>Simulate data under a tree model</p>
<pre class="r"><code>set.seed(1)
n = 40
x = cbind(c(rep(1,n),rep(0,n)), c(rep(0,n),rep(1,n)), c(rep(1,n/2),rep(0,3*n/2)), c(rep(0,n/2), rep(1,n/2), rep(0,n)), c(rep(0,n),rep(1,n/2),rep(0,n/2)), c(rep(0,3*n/2),rep(1,n/2)))
E = matrix(0.1*rexp(2*n*2*n),nrow=2*n)
E = E+t(E) #symmetric errors
S = x %*% diag(c(1,1,1,1,1,1)) %*% t(x) + E
image(S)</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Try initializing</p>
<pre class="r"><code>fit = snmu_init(S,1)
snmu_loglik(fit)</code></pre>
<pre><code>[1] -6496.862</code></pre>
<pre class="r"><code>fit = snmu_init(S,3)
snmu_loglik(fit)</code></pre>
<pre><code>[1] -6496.862</code></pre>
<p>Try updating b</p>
<pre class="r"><code>fit = snmu_init(S,1)
fit = snmu_update_b(fit)</code></pre>
<p>Try updating v</p>
<pre class="r"><code>fit = snmu_init(S,1)
fit = snmu_update_v(fit,ebnm_point_exponential)</code></pre>
<p>Here I try just fitting one factor</p>
<pre class="r"><code>set.seed(1)
fit = snmu_init(S,1)
for(i in 1:100){
  fit = snmu_update_v(fit,ebnm_point_exponential)
  fit = snmu_update_sigma_and_lambda_mom(fit)
  #fit = snmu_update_sigma_mle(fit)
}
plot(fit$v)</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now iteratively add factors. It eventually gets them all, but it adds
a null factor in the middle. I believe that is an initialization issue.
Note that I simply truncate the factor after a power update, without
checking on it’s “orientation” (+ or -). We might want to try making
sure it is correctly oriented before truncating (eg by checking v’Sv for
each orientation after truncation and choosing the orientation that
maximizes this?)</p>
<pre class="r"><code>for(i in 1:8){
  fit = snmu_add_new_factor(fit)
  k=ncol(fit$v)
  for(i in 1:100){
    # fit = snmu_update_b(fit) # I no longer bother with b for reasons noted in introduction
    fit = snmu_update_v(fit,ebnm_point_exponential,k)
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
    #fit = snmu_update_sigma_mle(fit)
  }
}
par(mfcol=c(3,3))
for(i in 1:ncol(fit$v)){
  plot(fit$v[,i])
}</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Try starting with a fixed intercept factor, and adding factors in a
greedy way. This one illustrates the problem that sometimes the point
exponential can find intermediate factors that capture more than one
branch. I should investigate this more.</p>
<pre class="r"><code>set.seed(4)
fit = snmu_init(S,1)
fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:8){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Try the same, but stopping after 6 factors to see what is going
on.</p>
<pre class="r"><code>set.seed(4)
fit = snmu_init(S,1)
fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:6){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Look at the value of d and the residuals;</p>
<pre class="r"><code>fit$d</code></pre>
<pre><code>[1] 17.06067 51.90636  0.00000 39.90091 19.81905 20.04218 13.58188</code></pre>
<pre class="r"><code>resid = S-vdvt(fit$v,fit$d)
image(resid)</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(resid[80,])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-14-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Try fitting more d iterations - it doesn’t change anything.</p>
<pre class="r"><code>for(i in 1:100){
  fit = snmu_update_d(fit)
}
fit$d</code></pre>
<pre><code>[1] 17.06067 51.90636  0.00000 39.90091 19.81905 20.04218 13.58188</code></pre>
<p>Interestingly, after only 10 iterations the fit is looking like a
single population. So it is not an initialization issue. It moves away
from this solution to the mixed solution. Possibly this is because the
second factor is slightly uneven between the two populations?</p>
<pre class="r"><code>fit = snmu_add_new_factor(fit)
plot(fit$v[,8])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for(i in 1:10){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
} 
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-16-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>After another 90 iterations we clearly see the mixed factor:</p>
<pre class="r"><code>for(i in 1:90){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
} 
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Here I try running the binormal after the point exponential.</p>
<pre class="r"><code>set.seed(4)
fit = snmu_init(S,1)
fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:6){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_binormal,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The binormal cleans up all the factors (as expected). Now adding a
new factor, and just using the point exponential actually picks out the
single population. This is consistent with the idea that the slightly
uneven second factor was driving it to find the mixed solution
previously. In any case this illustrates the delicacy of the solution,
and also suggests maybe running binormal after point exponential may be
a general helpful strategy.</p>
<pre class="r"><code>fit = snmu_add_new_factor(fit)
plot(fit$v[,8])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for(i in 1:1000){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
} 
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-19-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Here I try running the GB after the point exponential. It is a bit
less binary than the binormal prior, but has the same qualitative
effect.</p>
<pre class="r"><code>set.seed(4)
fit = snmu_init(S,1)
fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:6){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_generalized_binary,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}

fit = snmu_add_new_factor(fit)
for(i in 1:1000){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
} 
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="tree-with-strong-diagonal" class="section level3">
<h3>Tree with strong diagonal</h3>
<p>Here I simulate a tree but with a stronger diagonal. The point
exponenential now misses the top branches.</p>
<pre class="r"><code>S = x %*% diag(c(1,1,8,8,8,8)) %*% t(x) + E
image(S)</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>set.seed(4)
fit = snmu_init(S,1)

fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:8){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-21-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Try with point-exponential followed by binormal. It produces a
qualitively tree-like fit, although one factor is slightly trimodal.
(More iterations did not change this; not shown.)</p>
<pre class="r"><code>S = x %*% diag(c(1,1,8,8,8,8)) %*% t(x) + E
set.seed(4)
fit = snmu_init(S,1)

fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:8){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_binormal,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Same with GB instead of binormal. This one does not find the top
branches, maybe because it produces even more tri-modal factors that
capture the top branches. This illustrates that the binormal prior may
produce more binary factors than than the GB prior.</p>
<pre class="r"><code>S = x %*% diag(c(1,1,8,8,8,8)) %*% t(x) + E
set.seed(4)
fit = snmu_init(S,1)

fit$v = cbind(trunc_and_norm(rep(1,nrow(S))))
fit = snmu_update_d(fit)

for(k in 1:8){
  fit = snmu_add_new_factor(fit)
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_point_exponential,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
  for(i in 1:100){
    fit = snmu_update_v(fit,ebnm_generalized_binary,ncol(fit$v))
    fit = snmu_update_d(fit)
    fit = snmu_update_sigma_and_lambda_mom(fit)
  }  
}
par(mfcol=c(3,3))
for(k in 1:ncol(fit$v))
  plot(fit$v[,k])</code></pre>
<p><img src="figure/eb_snmu.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] ebnm_1.1-2     emg_1.0.9      moments_0.14.1

loaded via a namespace (and not attached):
 [1] sass_0.4.9        generics_0.1.3    ashr_2.2-63       stringi_1.8.4    
 [5] lattice_0.22-6    digest_0.6.37     magrittr_2.0.3    evaluate_1.0.3   
 [9] grid_4.4.2        fastmap_1.2.0     rprojroot_2.0.4   workflowr_1.7.1  
[13] jsonlite_1.8.9    Matrix_1.7-2      whisker_0.4.1     mixsqp_0.3-54    
[17] promises_1.3.2    scales_1.3.0      truncnorm_1.0-9   invgamma_1.1     
[21] jquerylib_0.1.4   cli_3.6.3         rlang_1.1.5       deconvolveR_1.2-1
[25] munsell_0.5.1     splines_4.4.2     cachem_1.1.0      yaml_2.3.10      
[29] tools_4.4.2       SQUAREM_2021.1    dplyr_1.1.4       colorspace_2.1-1 
[33] ggplot2_3.5.1     httpuv_1.6.15     vctrs_0.6.5       R6_2.5.1         
[37] lifecycle_1.0.4   git2r_0.35.0      stringr_1.5.1     fs_1.6.5         
[41] trust_0.1-8       irlba_2.3.5.1     pkgconfig_2.0.3   gtable_0.3.6     
[45] pillar_1.10.1     bslib_0.9.0       later_1.4.1       glue_1.8.0       
[49] Rcpp_1.0.14       xfun_0.50         tibble_3.2.1      tidyselect_1.2.1 
[53] rstudioapi_0.17.1 knitr_1.49        htmltools_0.5.8.1 rmarkdown_2.29   
[57] compiler_4.4.2    horseshoe_0.2.0  </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
