<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Matthew Stephens" />

<meta name="date" content="2025-11-11" />

<title>flash_pca</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephens999/misc">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">flash_pca</h1>
<h4 class="author">Matthew Stephens</h4>
<h4 class="date">2025-11-11</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-11-11
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree1eba385a76e30c27bff4518968b874bb1746cf87targetblank1eba385a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephens999/misc/tree/1eba385a76e30c27bff4518968b874bb1746cf87" target="_blank">1eba385</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree1eba385a76e30c27bff4518968b874bb1746cf87targetblank1eba385a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephens999/misc/tree/1eba385a76e30c27bff4518968b874bb1746cf87" target="_blank">1eba385</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/ALStruct_cache/
    Ignored:    data/.Rhistory
    Ignored:    data/methylation-data-for-matthew.rds
    Ignored:    data/pbmc/
    Ignored:    data/pbmc_purified.RData

Untracked files:
    Untracked:  .dropbox
    Untracked:  Icon
    Untracked:  analysis/GHstan.Rmd
    Untracked:  analysis/GTEX-cogaps.Rmd
    Untracked:  analysis/PACS.Rmd
    Untracked:  analysis/Rplot.png
    Untracked:  analysis/SPCAvRP.rmd
    Untracked:  analysis/abf_comparisons.Rmd
    Untracked:  analysis/admm_02.Rmd
    Untracked:  analysis/admm_03.Rmd
    Untracked:  analysis/bispca.Rmd
    Untracked:  analysis/cache/
    Untracked:  analysis/cholesky.Rmd
    Untracked:  analysis/compare-transformed-models.Rmd
    Untracked:  analysis/cormotif.Rmd
    Untracked:  analysis/cp_ash.Rmd
    Untracked:  analysis/eQTL.perm.rand.pdf
    Untracked:  analysis/eb_power2.Rmd
    Untracked:  analysis/eb_prepilot.Rmd
    Untracked:  analysis/eb_var.Rmd
    Untracked:  analysis/ebpmf1.Rmd
    Untracked:  analysis/ebpmf_sla_text.Rmd
    Untracked:  analysis/ebspca_sims.Rmd
    Untracked:  analysis/explore_psvd.Rmd
    Untracked:  analysis/fa_check_identify.Rmd
    Untracked:  analysis/fa_iterative.Rmd
    Untracked:  analysis/flash_cov_overlapping_groups_init.Rmd
    Untracked:  analysis/flash_test_tree.Rmd
    Untracked:  analysis/flashier_newgroups.Rmd
    Untracked:  analysis/flashier_nmf_triples.Rmd
    Untracked:  analysis/flashier_pbmc.Rmd
    Untracked:  analysis/flashier_snn_shifted_prior.Rmd
    Untracked:  analysis/greedy_ebpmf_exploration_00.Rmd
    Untracked:  analysis/ieQTL.perm.rand.pdf
    Untracked:  analysis/lasso_em_03.Rmd
    Untracked:  analysis/m6amash.Rmd
    Untracked:  analysis/mash_bhat_z.Rmd
    Untracked:  analysis/mash_ieqtl_permutations.Rmd
    Untracked:  analysis/methylation_example.Rmd
    Untracked:  analysis/mixsqp.Rmd
    Untracked:  analysis/mr.ash_lasso_init.Rmd
    Untracked:  analysis/mr.mash.test.Rmd
    Untracked:  analysis/mr_ash_modular.Rmd
    Untracked:  analysis/mr_ash_parameterization.Rmd
    Untracked:  analysis/mr_ash_ridge.Rmd
    Untracked:  analysis/mv_gaussian_message_passing.Rmd
    Untracked:  analysis/nejm.Rmd
    Untracked:  analysis/nmf_bg.Rmd
    Untracked:  analysis/nonneg_underapprox.Rmd
    Untracked:  analysis/normal_conditional_on_r2.Rmd
    Untracked:  analysis/normalize.Rmd
    Untracked:  analysis/pbmc.Rmd
    Untracked:  analysis/pca_binary_weighted.Rmd
    Untracked:  analysis/pca_l1.Rmd
    Untracked:  analysis/poisson_nmf_approx.Rmd
    Untracked:  analysis/poisson_shrink.Rmd
    Untracked:  analysis/poisson_transform.Rmd
    Untracked:  analysis/qrnotes.txt
    Untracked:  analysis/ridge_iterative_02.Rmd
    Untracked:  analysis/ridge_iterative_splitting.Rmd
    Untracked:  analysis/samps/
    Untracked:  analysis/sc_bimodal.Rmd
    Untracked:  analysis/shrinkage_comparisons_changepoints.Rmd
    Untracked:  analysis/susie_cov.Rmd
    Untracked:  analysis/susie_en.Rmd
    Untracked:  analysis/susie_z_investigate.Rmd
    Untracked:  analysis/svd-timing.Rmd
    Untracked:  analysis/temp.RDS
    Untracked:  analysis/temp.Rmd
    Untracked:  analysis/test-figure/
    Untracked:  analysis/test.Rmd
    Untracked:  analysis/test.Rpres
    Untracked:  analysis/test.md
    Untracked:  analysis/test_qr.R
    Untracked:  analysis/test_sparse.Rmd
    Untracked:  analysis/tree_dist_top_eigenvector.Rmd
    Untracked:  analysis/z.txt
    Untracked:  code/coordinate_descent_symNMF.R
    Untracked:  code/multivariate_testfuncs.R
    Untracked:  code/rqb.hacked.R
    Untracked:  data/4matthew/
    Untracked:  data/4matthew2/
    Untracked:  data/E-MTAB-2805.processed.1/
    Untracked:  data/ENSG00000156738.Sim_Y2.RDS
    Untracked:  data/GDS5363_full.soft.gz
    Untracked:  data/GSE41265_allGenesTPM.txt
    Untracked:  data/Muscle_Skeletal.ACTN3.pm1Mb.RDS
    Untracked:  data/P.rds
    Untracked:  data/Thyroid.FMO2.pm1Mb.RDS
    Untracked:  data/bmass.HaemgenRBC2016.MAF01.Vs2.MergedDataSources.200kRanSubset.ChrBPMAFMarkerZScores.vs1.txt.gz
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.NewSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.PreviousSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/eb_prepilot/
    Untracked:  data/finemap_data/fmo2.sim/b.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/data
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.ld
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.z
    Untracked:  data/finemap_data/fmo2.sim/pos.txt
    Untracked:  data/logm.csv
    Untracked:  data/m.cd.RDS
    Untracked:  data/m.cdu.old.RDS
    Untracked:  data/m.new.cd.RDS
    Untracked:  data/m.old.cd.RDS
    Untracked:  data/mainbib.bib.old
    Untracked:  data/mat.csv
    Untracked:  data/mat.txt
    Untracked:  data/mat_new.csv
    Untracked:  data/matrix_lik.rds
    Untracked:  data/paintor_data/
    Untracked:  data/running_data_chris.csv
    Untracked:  data/running_data_matthew.csv
    Untracked:  data/temp.txt
    Untracked:  data/y.txt
    Untracked:  data/y_f.txt
    Untracked:  data/zscore_jointLCLs_m6AQTLs_susie_eQTLpruned.rds
    Untracked:  data/zscore_jointLCLs_random.rds
    Untracked:  explore_udi.R
    Untracked:  output/fit.k10.rds
    Untracked:  output/fit.nn.pbmc.purified.rds
    Untracked:  output/fit.nn.rds
    Untracked:  output/fit.nn.s.001.rds
    Untracked:  output/fit.nn.s.01.rds
    Untracked:  output/fit.nn.s.1.rds
    Untracked:  output/fit.nn.s.10.rds
    Untracked:  output/fit.snn.s.001.rds
    Untracked:  output/fit.snn.s.01.nninit.rds
    Untracked:  output/fit.snn.s.01.rds
    Untracked:  output/fit.varbvs.RDS
    Untracked:  output/fit2.nn.pbmc.purified.rds
    Untracked:  output/glmnet.fit.RDS
    Untracked:  output/snn07.txt
    Untracked:  output/snn34.txt
    Untracked:  output/test.bv.txt
    Untracked:  output/test.gamma.txt
    Untracked:  output/test.hyp.txt
    Untracked:  output/test.log.txt
    Untracked:  output/test.param.txt
    Untracked:  output/test2.bv.txt
    Untracked:  output/test2.gamma.txt
    Untracked:  output/test2.hyp.txt
    Untracked:  output/test2.log.txt
    Untracked:  output/test2.param.txt
    Untracked:  output/test3.bv.txt
    Untracked:  output/test3.gamma.txt
    Untracked:  output/test3.hyp.txt
    Untracked:  output/test3.log.txt
    Untracked:  output/test3.param.txt
    Untracked:  output/test4.bv.txt
    Untracked:  output/test4.gamma.txt
    Untracked:  output/test4.hyp.txt
    Untracked:  output/test4.log.txt
    Untracked:  output/test4.param.txt
    Untracked:  output/test5.bv.txt
    Untracked:  output/test5.gamma.txt
    Untracked:  output/test5.hyp.txt
    Untracked:  output/test5.log.txt
    Untracked:  output/test5.param.txt

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/eb_snmu.Rmd
    Modified:   analysis/ebnm_binormal.Rmd
    Modified:   analysis/ebpower.Rmd
    Modified:   analysis/flashier_log1p.Rmd
    Modified:   analysis/flashier_sla_text.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/logistic_z_scores.Rmd
    Modified:   analysis/mr_ash_pen.Rmd
    Modified:   analysis/nmu_em.Rmd
    Modified:   analysis/susie_flash.Rmd
    Modified:   analysis/tap_free_energy.Rmd
    Modified:   misc.Rproj

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/flash_pca.Rmd</code>) and HTML
(<code>docs/flash_pca.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephens999/misc/blob/1eba385a76e30c27bff4518968b874bb1746cf87/analysis/flash_pca.Rmd" target="_blank">1eba385</a>
</td>
<td>
Matthew Stephens
</td>
<td>
2025-11-11
</td>
<td>
workflowr::wflow_publish("flash_pca.rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(flashier)</code></pre>
<pre><code>Loading required package: ebnm</code></pre>
<pre class="r"><code>library(fastICA)
library(PMA)</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction:</h2>
<p>Motivated by some results from <a href="fastica_02.html">ICA</a> I
want to try fitting <span class="math inline">\(X = M + lf&#39; +
E\)</span> where M is low rank.</p>
<p>Given <span class="math inline">\(l\)</span> and <span
class="math inline">\(f\)</span>, the maximum likelihood estimate for
<span class="math inline">\(M\)</span> is given by truncated SVD of
<span class="math inline">\(X-lf&#39;\)</span>, so I can iterate between
estimating <span class="math inline">\(l,f\)</span> and estimating <span
class="math inline">\(M\)</span>. This approach shoudl work for either
using flashier or PMD to estimate <span
class="math inline">\(l,f\)</span>.</p>
</div>
<div id="simulate-data" class="section level2">
<h2>Simulate data</h2>
<p>These are the same simulations as in (fastica_01.html)</p>
<pre class="r"><code>M &lt;- 10000 # Number of variants/samples (rows)
L &lt;- 10    # True number of latent factors
T &lt;- 100   # Number of traits/phenotypes (columns)

s_1 &lt;- 1   # Standard Deviation 1 (Spike component)
s_2 &lt;- 5   # Standard Deviation 2 (Slab component)
eps &lt;- 1e-2 # Standard Deviation for observation noise 

# Set seed for reproducibility
set.seed(42)

# Data Simulation (G = X %*% Y + noise)

# 3.1. Generating Standard Deviation Matrices (a and b)
# Elements are sampled from {s_1, s_2} [1, 2].
sd_choices &lt;- c(s_1, s_2)

# Matrix &#39;a&#39; (M x L): Standard deviations for X (Probabilities p=[0.7, 0.3]) [4]
p_a &lt;- c(0.7, 0.3)
a_vector &lt;- sample(sd_choices, size = M * L, replace = TRUE, prob = p_a)
a &lt;- matrix(a_vector, nrow = M, ncol = L)

# Matrix &#39;b&#39; (L x T): Standard deviations for Y (Probabilities p=[0.8, 0.2]) [4]
p_b &lt;- c(0.8, 0.2)
b_vector &lt;- sample(sd_choices, size = L * T, replace = TRUE, prob = p_b)
b &lt;- matrix(b_vector, nrow = L, ncol = T)

# Generating Latent Factors (X and Y)
# X is drawn from Normal(0, a)
X &lt;- matrix(rnorm(M * L, mean = 0, sd = a), nrow = M, ncol = L)

# Y is drawn from Normal(0, b)
Y &lt;- matrix(rnorm(L * T, mean = 0, sd = b), nrow = L, ncol = T)

# Generating Noise and Final Data Matrix G
# Noise is generated from Normal(0, eps)
noise &lt;- matrix(rnorm(M * T, mean = 0, sd = eps), nrow = M, ncol = T)

# Calculate the final data matrix G = X @ Y + noise
G &lt;- X %*% Y + noise</code></pre>
</div>
<div id="iterative-flash-with-pca" class="section level1">
<h1>Iterative flash with PCA</h1>
<p>Here I try the iterative approach with rank of M=9. I find that it
chooses a laplace prior with all its weight on non-null component, and
does not find a great solution.</p>
<pre class="r"><code>fit.fl = flash(G, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.</code></pre>
<pre class="r"><code>for(i in 1:5){
  M.svd = svd(G - fitted(fit.fl) )
  M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
  #fit.fl = flash(G-M, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)
  fit.fl = flash_update_data(fit.fl,G-M)
  fit.fl = flash_backfit(fit.fl)
  print(fit.fl$elbo)
}</code></pre>
<pre><code>Backfitting 1 factors (tolerance: 1.49e-02)...
  Difference between iterations is within 1.0e+06...
  Difference between iterations is within 1.0e+05...
  Difference between iterations is within 1.0e+04...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
Wrapping up...
Done.
[1] 3111599
Backfitting 1 factors (tolerance: 1.49e-02)...
  Difference between iterations is within 1.0e+01...
Wrapping up...
Done.
[1] 3141269
Backfitting 1 factors (tolerance: 1.49e-02)...
Wrapping up...
Done.
[1] 3141269
Backfitting 1 factors (tolerance: 1.49e-02)...
Wrapping up...
Done.
[1] 3141269
Backfitting 1 factors (tolerance: 1.49e-02)...
Wrapping up...
Done.
[1] 3141269</code></pre>
<pre class="r"><code>cor(X, fit.fl$L_pm)</code></pre>
<pre><code>             [,1]
 [1,] -0.38113764
 [2,]  0.18419586
 [3,] -0.08593986
 [4,]  0.13029310
 [5,] -0.17712154
 [6,]  0.56054887
 [7,] -0.11792697
 [8,] -0.61798516
 [9,]  0.25148413
[10,]  0.05945271</code></pre>
<pre class="r"><code>fit.fl$L_ghat</code></pre>
<pre><code>[[1]]
$pi
[1] 1.806438e-07 9.999998e-01

$mean
[1] 0 0

$scale
[1] 0.0000000 0.7150248

attr(,&quot;class&quot;)
[1] &quot;laplacemix&quot;
attr(,&quot;row.names&quot;)
[1] 1 2</code></pre>
<pre class="r"><code>fit.fl$elbo</code></pre>
<pre><code>[1] 3141269</code></pre>
<p>Try a normal mixture prior instead:</p>
<pre class="r"><code>fit.fl = flash(G, ebnm_fn = ebnm_normal_scale_mixture, greedy_Kmax = 1)</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.</code></pre>
<pre class="r"><code>cor(X, fit.fl$L_pm)</code></pre>
<pre><code>             [,1]
 [1,] -0.38140081
 [2,]  0.18573512
 [3,] -0.08625343
 [4,]  0.13224256
 [5,] -0.17841814
 [6,]  0.55913558
 [7,] -0.11852873
 [8,] -0.61631534
 [9,]  0.25340468
[10,]  0.06051061</code></pre>
<pre class="r"><code>for(i in 1:5){
  M.svd = svd(G - fitted(fit.fl) )
  M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
  #fit.fl = flash(G-M, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)
  fit.fl = flash_update_data(fit.fl,G-M)
  fit.fl = flash_backfit(fit.fl)
  print(fit.fl$elbo)
}</code></pre>
<pre><code>Backfitting 1 factors (tolerance: 1.49e-02)...
  Difference between iterations is within 1.0e+06...
  Difference between iterations is within 1.0e+05...
  Difference between iterations is within 1.0e+04...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
Wrapping up...
Done.
[1] 3131196
Backfitting 1 factors (tolerance: 1.49e-02)...
  Difference between iterations is within 1.0e+00...
Wrapping up...
Done.
[1] 3141348
Backfitting 1 factors (tolerance: 1.49e-02)...
Wrapping up...
Done.
[1] 3141348
Backfitting 1 factors (tolerance: 1.49e-02)...
Wrapping up...
Done.
[1] 3141348
Backfitting 1 factors (tolerance: 1.49e-02)...
Wrapping up...
Done.
[1] 3141348</code></pre>
<pre class="r"><code>cor(X, fit.fl$L_pm)</code></pre>
<pre><code>            [,1]
 [1,] -0.3815060
 [2,]  0.1857550
 [3,] -0.0862240
 [4,]  0.1322642
 [5,] -0.1784584
 [6,]  0.5592986
 [7,] -0.1185320
 [8,] -0.6164988
 [9,]  0.2534772
[10,]  0.0605242</code></pre>
<pre class="r"><code>fit.fl$L_ghat</code></pre>
<pre><code>[[1]]
$pi
 [1] 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.18094444
 [7] 0.03758443 0.00000000 0.21533611 0.02070509 0.00000000 0.41790590
[13] 0.07271325 0.05481079 0.00000000 0.00000000 0.00000000 0.00000000
[19] 0.00000000 0.00000000 0.00000000 0.00000000

$mean
 [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

$sd
 [1] 0.0000000 0.1229333 0.1891227 0.2529762 0.3202798 0.3940946 0.4768439
 [8] 0.5708588 0.6786031 0.8028042 0.9465530 1.1133963 1.3074330 1.5334192
[15] 1.7968879 2.1042857 2.4631315 2.8822015 3.3717434 3.9437279 4.6121413
[22] 5.3933273

attr(,&quot;class&quot;)
[1] &quot;normalmix&quot;
attr(,&quot;row.names&quot;)
 [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22</code></pre>
<pre class="r"><code>fit.fl$elbo</code></pre>
<pre><code>[1] 3141348</code></pre>
<p>Now try with initializing at an ica solution. It still uses a laplace
prior, but it essentially converges to the ica solution (ie does not
move much) and the elbo is better</p>
<pre class="r"><code>fit.ica = fastICA(G, n.comp = 10)
s = fit.ica$S[,1]
a = fit.ica$A[1,]
M.svd = svd(G - s %*% t(a) )
M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
fit.fl2 = flash(G-M, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.</code></pre>
<pre class="r"><code>fit.fl2$L_ghat</code></pre>
<pre><code>[[1]]
$pi
[1] 1.940483e-06 9.999981e-01

$mean
[1] 0 0

$scale
[1] 0.0000000 0.4334429

attr(,&quot;class&quot;)
[1] &quot;laplacemix&quot;
attr(,&quot;row.names&quot;)
[1] 1 2</code></pre>
<pre class="r"><code>fit.fl2$elbo</code></pre>
<pre><code>[1] 3148073</code></pre>
<pre class="r"><code>for(i in 1:5){
  M.svd = svd(G - fitted(fit.fl2) )
  M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
  fit.fl2 = flash(G-M, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)
  print(fit.fl$elbo)
}</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 3141348
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 3141348
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 3141348
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 3141348
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 3141348</code></pre>
<pre class="r"><code>cor(X, fit.fl2$L_pm)</code></pre>
<pre><code>               [,1]
 [1,] -2.211661e-03
 [2,] -2.436869e-03
 [3,]  2.869316e-05
 [4,]  7.598077e-03
 [5,]  1.470651e-03
 [6,] -3.292375e-03
 [7,] -9.997270e-01
 [8,] -3.886822e-03
 [9,]  8.330709e-03
[10,] -4.525288e-03</code></pre>
<pre class="r"><code>fit.fl2$L_ghat</code></pre>
<pre><code>[[1]]
$pi
[1] 1.685282e-06 9.999983e-01

$mean
[1] 0 0

$scale
[1] 0.0000000 0.4334428

attr(,&quot;class&quot;)
[1] &quot;laplacemix&quot;
attr(,&quot;row.names&quot;)
[1] 1 2</code></pre>
<pre class="r"><code>fit.fl2$F_ghat</code></pre>
<pre><code>[[1]]
$pi
[1] 6.48661e-11 1.00000e+00

$mean
[1] 0 0

$scale
[1] 0.000000 4.767295

attr(,&quot;class&quot;)
[1] &quot;laplacemix&quot;
attr(,&quot;row.names&quot;)
[1] 1 2</code></pre>
<pre class="r"><code>fit.fl2$elbo</code></pre>
<pre><code>[1] 3149452</code></pre>
<pre class="r"><code>fit.fl$elbo</code></pre>
<pre><code>[1] 3141348</code></pre>
<p>Here I try the same thing, but using just the first 8 PCs in an
attempt to provide a bit more “wiggle room” for the optimization to
move. But I found convergence is very slow here.</p>
<pre class="r"><code>fit.fl = flash(G, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.</code></pre>
<pre class="r"><code>for(i in 1:50){
  M.svd = svd(G - fitted(fit.fl) )
  M = M.svd$u[,1:8] %*% (M.svd$d[1:8] * t(M.svd$v[,1:8]))
  fit.fl = flash(G-M, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)
  print(fit.fl$elbo)
}</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740099
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740098
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740098
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740097
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740097
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740096
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740096
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740096
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740095
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740095
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740094
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740094
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740093
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740093
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740093
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740092
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740092
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740092
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740091
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740091
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740090
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740090
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740090
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740089
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740089
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740089
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740088
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740088
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740088
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740087
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740087
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740087
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740086
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740086
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740086
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740085
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740085
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740085
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740085
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740084
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740084
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740084
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740083
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740083
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740083
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740082
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740082
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740082
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740081
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] -2740081</code></pre>
<pre class="r"><code>cor(X, fit.fl$L_pm)</code></pre>
<pre><code>             [,1]
 [1,] -0.37800842
 [2,]  0.17910674
 [3,] -0.08737775
 [4,]  0.12771926
 [5,] -0.17455302
 [6,]  0.56250455
 [7,] -0.11533349
 [8,] -0.62305161
 [9,]  0.24789657
[10,]  0.05560268</code></pre>
<pre class="r"><code>fit.fl$L_ghat</code></pre>
<pre><code>[[1]]
$pi
[1] 1.131635e-06 9.999989e-01

$mean
[1] 0 0

$scale
[1] 0.0000000 0.7134562

attr(,&quot;class&quot;)
[1] &quot;laplacemix&quot;
attr(,&quot;row.names&quot;)
[1] 1 2</code></pre>
<pre class="r"><code>fit.fl$elbo</code></pre>
<pre><code>[1] -2740081</code></pre>
<p>Try nuclear norm regularization on M rather than hard rank
constraint.</p>
<pre class="r"><code>fit.fl = flash(G, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.</code></pre>
<pre class="r"><code>for(i in 1:50){
  M.svd = svd(G - fitted(fit.fl) )
  M.svd$d = pmax(0, M.svd$d - 10)
  M = M.svd$u %*% (M.svd$d * t(M.svd$v))
  fit.fl = flash(G-M, ebnm_fn = ebnm_point_laplace, greedy_Kmax = 1)
  print(fit.fl$elbo)
}</code></pre>
<pre><code>Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1924275
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1923715
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1936740
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1946657
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1951180
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1953225
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954144
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954551
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954727
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954799
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954824
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954827
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954821
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954810
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954797
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954783
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954769
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954755
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954740
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954726
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954712
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954697
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954683
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954669
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954655
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954640
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954626
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954612
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954598
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954584
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954569
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954555
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954541
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954527
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954513
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954499
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954485
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954471
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954457
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954443
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954429
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954415
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954401
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954387
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954373
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954359
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954345
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954330
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954315
Adding factor 1 to flash object...
Wrapping up...
Done.
Nullchecking 1 factors...
Done.
[1] 1954306</code></pre>
<pre class="r"><code>cor(X, fit.fl$L_pm)</code></pre>
<pre><code>             [,1]
 [1,] -0.38122318
 [2,]  0.18435429
 [3,] -0.08616472
 [4,]  0.13039632
 [5,] -0.17723444
 [6,]  0.56045909
 [7,] -0.11810855
 [8,] -0.61778328
 [9,]  0.25160371
[10,]  0.05952674</code></pre>
<pre class="r"><code>fit.fl$L_ghat</code></pre>
<pre><code>[[1]]
$pi
[1] 2.814814e-07 9.999997e-01

$mean
[1] 0 0

$scale
[1] 0.000000 0.715645

attr(,&quot;class&quot;)
[1] &quot;laplacemix&quot;
attr(,&quot;row.names&quot;)
[1] 1 2</code></pre>
<pre class="r"><code>fit.fl$elbo</code></pre>
<pre><code>[1] 1954306</code></pre>
<div id="iterative-pmd-with-pca" class="section level2">
<h2>Iterative PMD with PCA</h2>
<p>Here I try the same with PMD</p>
<pre class="r"><code>library(PMA)
fit.pmd = PMD(G,&quot;standard&quot;,sumabs=0.5,center=F)</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1234567891011121314151617181920</code></pre>
<pre class="r"><code>cor(X, fit.pmd$u)</code></pre>
<pre><code>             [,1]
 [1,]  0.26840307
 [2,] -0.08054940
 [3,]  0.06713861
 [4,] -0.01299302
 [5,]  0.10758749
 [6,] -0.36451223
 [7,]  0.05598098
 [8,]  0.81813865
 [9,] -0.12268001
[10,] -0.07062926</code></pre>
<pre class="r"><code>for(i in 1:10){
  M.svd = svd(G - fit.pmd$u %*% t(fit.pmd$d * fit.pmd$v) )
  M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
  fit.pmd = PMD(G-M, v = fit.pmd$v, type=&quot;standard&quot;, sumabs= 0.5, center=F,niter=1)
}</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre class="r"><code>cor(X, fit.pmd$u)</code></pre>
<pre><code>              [,1]
 [1,]  0.011377423
 [2,] -0.011759704
 [3,] -0.008862186
 [4,]  0.011258508
 [5,]  0.016953914
 [6,] -0.055848148
 [7,]  0.011011700
 [8,]  0.986368084
 [9,] -0.014366883
[10,]  0.011987536</code></pre>
<pre class="r"><code># fit term
sum((G-M- fit.pmd$u %*% (fit.pmd$d * t(fit.pmd$v)))^2) </code></pre>
<pre><code>[1] 2029164</code></pre>
<pre class="r"><code># penalty term (seems to be fixed)
sum(abs(fit.pmd$u)) * fit.pmd$sumabsu + sum(abs(fit.pmd$v)) * fit.pmd$sumabsv</code></pre>
<pre><code>[1] 2525</code></pre>
<p>This worked well. Now I’ll try a lesser penalty.</p>
<pre class="r"><code>fit.pmd = PMD(G,&quot;standard&quot;,sumabs=0.9,center=F)</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1</code></pre>
<pre class="r"><code>cor(X, fit.pmd$u)</code></pre>
<pre><code>             [,1]
 [1,]  0.38293438
 [2,] -0.19070578
 [3,]  0.08766478
 [4,] -0.13750981
 [5,]  0.18229298
 [6,] -0.55618217
 [7,]  0.12076554
 [8,]  0.61146048
 [9,] -0.25850008
[10,] -0.06340065</code></pre>
<pre class="r"><code>for(i in 1:10){
  M.svd = svd(G - fit.pmd$u %*% t(fit.pmd$d * fit.pmd$v) )
  M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
  fit.pmd = PMD(G-M, v = fit.pmd$v, type=&quot;standard&quot;, sumabs= 0.9, center=F,niter=1)
  print(sum((G-M- fit.pmd$u %*% (fit.pmd$d * t(fit.pmd$v)))^2) )
}</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre><code>Warning in PMDL1L1(x, sumabs = sumabs, sumabsu = sumabsu, sumabsv = sumabsv, :
PMDL1L1 was run without first subtracting out the mean of x.</code></pre>
<pre><code>1
[1] 89.97461</code></pre>
<pre class="r"><code>cor(X, fit.pmd$u)</code></pre>
<pre><code>             [,1]
 [1,]  0.38293438
 [2,] -0.19070578
 [3,]  0.08766478
 [4,] -0.13750981
 [5,]  0.18229298
 [6,] -0.55618217
 [7,]  0.12076554
 [8,]  0.61146048
 [9,] -0.25850008
[10,] -0.06340065</code></pre>
<pre class="r"><code># fit term
sum((G-M- fit.pmd$u %*% (fit.pmd$d * t(fit.pmd$v)))^2) </code></pre>
<pre><code>[1] 89.97461</code></pre>
<pre class="r"><code># penalty term (seems to be fixed)
sum(abs(fit.pmd$u)) * fit.pmd$sumabsu + sum(abs(fit.pmd$v)) * fit.pmd$sumabsv</code></pre>
<pre><code>[1] 6784.095</code></pre>
<p>Here I compare the fit and penalty at the fastICA solution to see if
this is a convergence issue. It was a bit of a suprise to see that the
ICA solution does not get close to the PMD fit term. This is presumably
because it does not optimize for mse. Also we see that the M.svd$d has 9
large eigenvalues, and one that is not large but bigger than background.
This is the one that ICA is not fitting. That is, ICA does not actually
leave a rank 9 matrix but actually a rank 10 matrix. Presumably this is
because ICA fits the top 10PCs not the actual data matrix, so it leaves
some of the “noise” unfitted, but maybe actually PMD is fitting more
noise? (Is this an advantage of an initial step doing PCA? It might be.
Maybe we should try PMD on the top PCs rather than the full data
matrix?)</p>
<pre class="r"><code>M.svd = svd(G - s %*% t(a) )
M = M.svd$u[,1:9] %*% (M.svd$d[1:9] * t(M.svd$v[,1:9]))
sum((G-M- s %*% t(a))^2)</code></pre>
<pre><code>[1] 93.62801</code></pre>
<pre class="r"><code>M.svd$d</code></pre>
<pre><code>  [1] 9176.2905200 8113.6850689 7919.9268198 7220.2153718 7120.6590558
  [6] 6467.8700988 6108.6111561 5314.9885635 3673.8448242    1.9145568
 [11]    1.0879917    1.0818923    1.0781103    1.0769687    1.0749752
 [16]    1.0737024    1.0705980    1.0673045    1.0639946    1.0618220
 [21]    1.0600228    1.0576188    1.0561401    1.0532369    1.0514490
 [26]    1.0486938    1.0472644    1.0449246    1.0417742    1.0408329
 [31]    1.0397812    1.0379476    1.0359387    1.0348096    1.0338034
 [36]    1.0318856    1.0307806    1.0301591    1.0260115    1.0235496
 [41]    1.0227009    1.0217301    1.0191786    1.0173757    1.0149219
 [46]    1.0140178    1.0121285    1.0109634    1.0095153    1.0065901
 [51]    1.0061486    1.0041182    1.0010232    0.9995755    0.9977064
 [56]    0.9975979    0.9957359    0.9946129    0.9941258    0.9918242
 [61]    0.9886653    0.9882669    0.9858695    0.9823833    0.9814767
 [66]    0.9807947    0.9802120    0.9787143    0.9766344    0.9738211
 [71]    0.9730331    0.9720059    0.9702359    0.9665484    0.9661569
 [76]    0.9645165    0.9629077    0.9604952    0.9594310    0.9568659
 [81]    0.9551573    0.9545256    0.9520256    0.9492639    0.9462706
 [86]    0.9448512    0.9440892    0.9428766    0.9397537    0.9378132
 [91]    0.9356203    0.9344767    0.9306737    0.9276352    0.9267261
 [96]    0.9249556    0.9239296    0.9199207    0.9158973    0.9126204</code></pre>
<pre class="r"><code>svd((G-fit.pmd$u %*% (fit.pmd$d * t(fit.pmd$v))))$d</code></pre>
<pre><code>  [1] 8.117499e+03 7.920111e+03 7.256598e+03 7.134314e+03 6.478365e+03
  [6] 6.148451e+03 5.338734e+03 4.748096e+03 3.608830e+03 1.088411e+00
 [11] 1.081894e+00 1.078265e+00 1.076970e+00 1.075021e+00 1.073817e+00
 [16] 1.070731e+00 1.067332e+00 1.064023e+00 1.061822e+00 1.060127e+00
 [21] 1.057620e+00 1.056216e+00 1.053259e+00 1.051613e+00 1.048720e+00
 [26] 1.047286e+00 1.044964e+00 1.042030e+00 1.040842e+00 1.039811e+00
 [31] 1.037961e+00 1.035977e+00 1.034845e+00 1.033894e+00 1.032034e+00
 [36] 1.030914e+00 1.030176e+00 1.026102e+00 1.023608e+00 1.022779e+00
 [41] 1.021776e+00 1.019187e+00 1.017391e+00 1.014994e+00 1.014120e+00
 [46] 1.012301e+00 1.010972e+00 1.009856e+00 1.006717e+00 1.006161e+00
 [51] 1.004120e+00 1.001031e+00 9.995807e-01 9.977976e-01 9.976091e-01
 [56] 9.957384e-01 9.946141e-01 9.942168e-01 9.921946e-01 9.888689e-01
 [61] 9.883603e-01 9.859798e-01 9.824715e-01 9.814767e-01 9.810739e-01
 [66] 9.803592e-01 9.787418e-01 9.766685e-01 9.738279e-01 9.730336e-01
 [71] 9.720156e-01 9.702887e-01 9.665486e-01 9.661805e-01 9.646655e-01
 [76] 9.629099e-01 9.605022e-01 9.594346e-01 9.568751e-01 9.551861e-01
 [81] 9.545383e-01 9.521384e-01 9.493240e-01 9.463089e-01 9.449444e-01
 [86] 9.440933e-01 9.430927e-01 9.397773e-01 9.378201e-01 9.356860e-01
 [91] 9.344981e-01 9.307157e-01 9.276357e-01 9.267522e-01 9.249570e-01
 [96] 9.240263e-01 9.199398e-01 9.159090e-01 9.126318e-01 1.617511e-11</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.6.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] PMA_1.2-4       fastICA_1.2-7   flashier_1.0.56 ebnm_1.1-34    

loaded via a namespace (and not attached):
 [1] softImpute_1.4-3     gtable_0.3.6         xfun_0.52           
 [4] bslib_0.9.0          ggplot2_3.5.2        htmlwidgets_1.6.4   
 [7] ggrepel_0.9.6        lattice_0.22-6       quadprog_1.5-8      
[10] vctrs_0.6.5          tools_4.4.2          generics_0.1.4      
[13] parallel_4.4.2       Polychrome_1.5.4     tibble_3.3.0        
[16] pkgconfig_2.0.3      Matrix_1.7-2         data.table_1.17.6   
[19] SQUAREM_2021.1       RColorBrewer_1.1-3   RcppParallel_5.1.10 
[22] scatterplot3d_0.3-44 lifecycle_1.0.4      truncnorm_1.0-9     
[25] compiler_4.4.2       farver_2.1.2         stringr_1.5.1       
[28] git2r_0.35.0         progress_1.2.3       RhpcBLASctl_0.23-42 
[31] httpuv_1.6.15        htmltools_0.5.8.1    sass_0.4.10         
[34] lazyeval_0.2.2       yaml_2.3.10          plotly_4.11.0       
[37] crayon_1.5.3         tidyr_1.3.1          later_1.4.2         
[40] pillar_1.10.2        jquerylib_0.1.4      whisker_0.4.1       
[43] uwot_0.2.3           cachem_1.1.0         trust_0.1-8         
[46] gtools_3.9.5         tidyselect_1.2.1     digest_0.6.37       
[49] Rtsne_0.17           stringi_1.8.7        purrr_1.0.4         
[52] dplyr_1.1.4          ashr_2.2-66          splines_4.4.2       
[55] cowplot_1.1.3        rprojroot_2.0.4      fastmap_1.2.0       
[58] grid_4.4.2           colorspace_2.1-1     cli_3.6.5           
[61] invgamma_1.1         magrittr_2.0.3       prettyunits_1.2.0   
[64] scales_1.4.0         promises_1.3.3       horseshoe_0.2.0     
[67] httr_1.4.7           rmarkdown_2.29       fastTopics_0.7-07   
[70] deconvolveR_1.2-1    workflowr_1.7.1      hms_1.1.3           
[73] pbapply_1.7-2        evaluate_1.0.4       knitr_1.50          
[76] viridisLite_0.4.2    irlba_2.3.5.1        rlang_1.1.6         
[79] Rcpp_1.0.14          mixsqp_0.3-54        glue_1.8.0          
[82] rstudioapi_0.17.1    jsonlite_2.0.0       R6_2.6.1            
[85] fs_1.6.6            </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
