<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Matthew Stephens" />

<meta name="date" content="2024-07-31" />

<title>flash_cov_overlapping_groups</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephens999/misc">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">flash_cov_overlapping_groups</h1>
<h4 class="author">Matthew Stephens</h4>
<h4 class="date">2024-07-31</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-09-14
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree2e5d38a5405c4926a0d075aceef524cdcb768897targetblank2e5d38aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephens999/misc/tree/2e5d38a5405c4926a0d075aceef524cdcb768897" target="_blank">2e5d38a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree2e5d38a5405c4926a0d075aceef524cdcb768897targetblank2e5d38aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephens999/misc/tree/2e5d38a5405c4926a0d075aceef524cdcb768897" target="_blank">2e5d38a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/ALStruct_cache/
    Ignored:    data/.Rhistory
    Ignored:    data/methylation-data-for-matthew.rds
    Ignored:    data/pbmc/
    Ignored:    data/pbmc_purified.RData

Untracked files:
    Untracked:  .dropbox
    Untracked:  Icon
    Untracked:  analysis/GHstan.Rmd
    Untracked:  analysis/GTEX-cogaps.Rmd
    Untracked:  analysis/PACS.Rmd
    Untracked:  analysis/Rplot.png
    Untracked:  analysis/SPCAvRP.rmd
    Untracked:  analysis/abf_comparisons.Rmd
    Untracked:  analysis/admm_02.Rmd
    Untracked:  analysis/admm_03.Rmd
    Untracked:  analysis/cache/
    Untracked:  analysis/cholesky.Rmd
    Untracked:  analysis/compare-transformed-models.Rmd
    Untracked:  analysis/cormotif.Rmd
    Untracked:  analysis/cp_ash.Rmd
    Untracked:  analysis/eQTL.perm.rand.pdf
    Untracked:  analysis/eb_prepilot.Rmd
    Untracked:  analysis/eb_var.Rmd
    Untracked:  analysis/ebpmf1.Rmd
    Untracked:  analysis/ebpmf_sla_text.Rmd
    Untracked:  analysis/ebspca_sims.Rmd
    Untracked:  analysis/explore_psvd.Rmd
    Untracked:  analysis/fa_check_identify.Rmd
    Untracked:  analysis/fa_iterative.Rmd
    Untracked:  analysis/flash_test_tree.Rmd
    Untracked:  analysis/flash_tree.Rmd
    Untracked:  analysis/flashier_newgroups.Rmd
    Untracked:  analysis/flashier_nmf_triples.Rmd
    Untracked:  analysis/flashier_pbmc.Rmd
    Untracked:  analysis/flashier_snn_shifted_prior.Rmd
    Untracked:  analysis/greedy_ebpmf_exploration_00.Rmd
    Untracked:  analysis/ieQTL.perm.rand.pdf
    Untracked:  analysis/lasso_em_03.Rmd
    Untracked:  analysis/m6amash.Rmd
    Untracked:  analysis/mash_bhat_z.Rmd
    Untracked:  analysis/mash_ieqtl_permutations.Rmd
    Untracked:  analysis/methylation_example.Rmd
    Untracked:  analysis/mixsqp.Rmd
    Untracked:  analysis/mr.ash_lasso_init.Rmd
    Untracked:  analysis/mr.mash.test.Rmd
    Untracked:  analysis/mr_ash_modular.Rmd
    Untracked:  analysis/mr_ash_parameterization.Rmd
    Untracked:  analysis/mr_ash_ridge.Rmd
    Untracked:  analysis/mv_gaussian_message_passing.Rmd
    Untracked:  analysis/nejm.Rmd
    Untracked:  analysis/nmf_bg.Rmd
    Untracked:  analysis/normal_conditional_on_r2.Rmd
    Untracked:  analysis/normalize.Rmd
    Untracked:  analysis/pbmc.Rmd
    Untracked:  analysis/pca_binary_weighted.Rmd
    Untracked:  analysis/pca_l1.Rmd
    Untracked:  analysis/poisson_nmf_approx.Rmd
    Untracked:  analysis/poisson_shrink.Rmd
    Untracked:  analysis/poisson_transform.Rmd
    Untracked:  analysis/qrnotes.txt
    Untracked:  analysis/ridge_iterative_02.Rmd
    Untracked:  analysis/ridge_iterative_splitting.Rmd
    Untracked:  analysis/samps/
    Untracked:  analysis/sc_bimodal.Rmd
    Untracked:  analysis/shrinkage_comparisons_changepoints.Rmd
    Untracked:  analysis/susie_cov.Rmd
    Untracked:  analysis/susie_en.Rmd
    Untracked:  analysis/susie_z_investigate.Rmd
    Untracked:  analysis/svd-timing.Rmd
    Untracked:  analysis/temp.RDS
    Untracked:  analysis/temp.Rmd
    Untracked:  analysis/test-figure/
    Untracked:  analysis/test.Rmd
    Untracked:  analysis/test.Rpres
    Untracked:  analysis/test.md
    Untracked:  analysis/test_qr.R
    Untracked:  analysis/test_sparse.Rmd
    Untracked:  analysis/tree_dist_top_eigenvector.Rmd
    Untracked:  analysis/z.txt
    Untracked:  code/multivariate_testfuncs.R
    Untracked:  code/rqb.hacked.R
    Untracked:  data/4matthew/
    Untracked:  data/4matthew2/
    Untracked:  data/E-MTAB-2805.processed.1/
    Untracked:  data/ENSG00000156738.Sim_Y2.RDS
    Untracked:  data/GDS5363_full.soft.gz
    Untracked:  data/GSE41265_allGenesTPM.txt
    Untracked:  data/Muscle_Skeletal.ACTN3.pm1Mb.RDS
    Untracked:  data/P.rds
    Untracked:  data/Thyroid.FMO2.pm1Mb.RDS
    Untracked:  data/bmass.HaemgenRBC2016.MAF01.Vs2.MergedDataSources.200kRanSubset.ChrBPMAFMarkerZScores.vs1.txt.gz
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.NewSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.PreviousSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/eb_prepilot/
    Untracked:  data/finemap_data/fmo2.sim/b.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/data
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.ld
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.z
    Untracked:  data/finemap_data/fmo2.sim/pos.txt
    Untracked:  data/logm.csv
    Untracked:  data/m.cd.RDS
    Untracked:  data/m.cdu.old.RDS
    Untracked:  data/m.new.cd.RDS
    Untracked:  data/m.old.cd.RDS
    Untracked:  data/mainbib.bib.old
    Untracked:  data/mat.csv
    Untracked:  data/mat.txt
    Untracked:  data/mat_new.csv
    Untracked:  data/matrix_lik.rds
    Untracked:  data/paintor_data/
    Untracked:  data/running_data_chris.csv
    Untracked:  data/running_data_matthew.csv
    Untracked:  data/temp.txt
    Untracked:  data/y.txt
    Untracked:  data/y_f.txt
    Untracked:  data/zscore_jointLCLs_m6AQTLs_susie_eQTLpruned.rds
    Untracked:  data/zscore_jointLCLs_random.rds
    Untracked:  explore_udi.R
    Untracked:  output/fit.k10.rds
    Untracked:  output/fit.nn.pbmc.purified.rds
    Untracked:  output/fit.nn.rds
    Untracked:  output/fit.nn.s.001.rds
    Untracked:  output/fit.nn.s.01.rds
    Untracked:  output/fit.nn.s.1.rds
    Untracked:  output/fit.nn.s.10.rds
    Untracked:  output/fit.snn.s.001.rds
    Untracked:  output/fit.snn.s.01.nninit.rds
    Untracked:  output/fit.snn.s.01.rds
    Untracked:  output/fit.varbvs.RDS
    Untracked:  output/fit2.nn.pbmc.purified.rds
    Untracked:  output/glmnet.fit.RDS
    Untracked:  output/snn07.txt
    Untracked:  output/snn34.txt
    Untracked:  output/test.bv.txt
    Untracked:  output/test.gamma.txt
    Untracked:  output/test.hyp.txt
    Untracked:  output/test.log.txt
    Untracked:  output/test.param.txt
    Untracked:  output/test2.bv.txt
    Untracked:  output/test2.gamma.txt
    Untracked:  output/test2.hyp.txt
    Untracked:  output/test2.log.txt
    Untracked:  output/test2.param.txt
    Untracked:  output/test3.bv.txt
    Untracked:  output/test3.gamma.txt
    Untracked:  output/test3.hyp.txt
    Untracked:  output/test3.log.txt
    Untracked:  output/test3.param.txt
    Untracked:  output/test4.bv.txt
    Untracked:  output/test4.gamma.txt
    Untracked:  output/test4.hyp.txt
    Untracked:  output/test4.log.txt
    Untracked:  output/test4.param.txt
    Untracked:  output/test5.bv.txt
    Untracked:  output/test5.gamma.txt
    Untracked:  output/test5.hyp.txt
    Untracked:  output/test5.log.txt
    Untracked:  output/test5.param.txt

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/flashier_log1p.Rmd
    Modified:   analysis/flashier_sla_text.Rmd
    Modified:   analysis/logistic_z_scores.Rmd
    Modified:   analysis/mr_ash_pen.Rmd
    Modified:   analysis/susie_flash.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/flash_cov_overlapping_groups.Rmd</code>) and HTML
(<code>docs/flash_cov_overlapping_groups.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephens999/misc/blob/2e5d38a5405c4926a0d075aceef524cdcb768897/analysis/flash_cov_overlapping_groups.Rmd" target="_blank">2e5d38a</a>
</td>
<td>
Matthew Stephens
</td>
<td>
2024-09-14
</td>
<td>
workflowr::wflow_publish("analysis/flash_cov_overlapping_groups.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>I wanted to examine how applying flash to gram matrix works in some
simple simulations involving overlapping groups (L is binary, F is
normal). I simulate binary factors, and use point-exponential priors.
The results are somewhat promising (after back-fitting). Even when it
doesn’t separate the groups, each factor captures a small number of
groups. It suggests one could maybe do some post-processing/follow up to
try to split these compound factors into multiple factors, one for each
group.</p>
</div>
<div id="three-overlapping-groups" class="section level2">
<h2>Three overlapping groups</h2>
<p>I simulate some data with 3 overlapping groups, each containing a
random 1/10 of the observations (so the groups do not overlap too much).
I start with very little noise.</p>
<pre class="r"><code>set.seed(1)
library(flashier)</code></pre>
<pre><code>Loading required package: ebnm</code></pre>
<pre class="r"><code>n = 100
p = 1000
k = 3
L= matrix(rbinom(k*n,1,0.1),nrow=n)
F = matrix(rnorm(k*p),nrow=p)
X = L %*% t(F) + rnorm(n*p,0,0.001)
plot(rowSums(L))</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>I fit with both greedy and backfitting:</p>
<pre class="r"><code>fit = flash(X %*% t(X), ebnm_fn = ebnm_point_exponential)</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...</code></pre>
<pre><code>Warning in scale.EF(EF): Fitting stopped after the initialization function
failed to find a non-zero factor.</code></pre>
<pre><code>Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Nullchecking 4 factors...
Done.</code></pre>
<pre class="r"><code>fit.bf = flash(X %*% t(X), ebnm_fn = ebnm_point_exponential, backfit =TRUE)</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...</code></pre>
<pre><code>Warning in scale.EF(EF): Fitting stopped after the initialization function
failed to find a non-zero factor.</code></pre>
<pre><code>Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 4 factors (tolerance: 1.49e-04)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
  Difference between iterations is within 1.0e-03...
  Difference between iterations is within 1.0e-04...
Wrapping up...
Done.
Nullchecking 4 factors...
Done.</code></pre>
<p>The greedy fit starts by initially picking out multiple groups (it is
strongly correlated with how many groups each sample is in):</p>
<pre class="r"><code>plot(rowSums(L),fit$L_pm[,1])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>But the backfitting helps fix this: now the estimated L is almost
binary and picks out the groups correctly (groups 2 and 3 are
switched):</p>
<pre class="r"><code>plot(rowSums(L), fit.bf$L_pm[,1])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(L[,1],fit.bf$L_pm[,1])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(L[,2],fit.bf$L_pm[,3])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-4-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(L[,3],fit.bf$L_pm[,2])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-4-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>Try point laplace to see what happens. In this case it basically
explains everything with 2 factors (3rd factor has small contribution)
and the first factor is capturing the number of groups - backfitting
does not work as with point exponential here.</p>
<pre class="r"><code>fit.pl.bf = flash(X %*% t(X), ebnm_fn = ebnm_point_laplace, backfit =TRUE)</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 3 factors (tolerance: 1.49e-04)...
  Difference between iterations is within 1.0e+04...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
Wrapping up...
Done.
Nullchecking 3 factors...
Done.</code></pre>
<pre class="r"><code>fit.pl.bf</code></pre>
<pre><code>Flash object with 3 factors.
  Proportion of variance explained:
    Factor 1: 0.618
    Factor 2: 0.322
    Factor 3: 0.061
  Variational lower bound: 68070.739</code></pre>
<pre class="r"><code>plot(rowSums(L),fit.pl.bf$L_pm[,1])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="overlapping-groups" class="section level2">
<h2>10 overlapping groups</h2>
<p>Now I try something more challenging with 10 groups:</p>
<pre class="r"><code>set.seed(1)
k=10
L= matrix(rbinom(k*n,1,0.1),nrow=n)
F = matrix(rnorm(k*p),nrow=p)
X = L %*% t(F) + rnorm(n*p,0,0.001)
plot(rowSums(L))</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Fit greedy and backfitting:</p>
<pre class="r"><code>fit = flash(X %*% t(X), ebnm_fn = ebnm_point_exponential)</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...</code></pre>
<pre><code>Warning in scale.EF(EF): Fitting stopped after the initialization function
failed to find a non-zero factor.</code></pre>
<pre><code>Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Nullchecking 6 factors...
Done.</code></pre>
<pre class="r"><code>fit.bf = flash(X %*% t(X), ebnm_fn = ebnm_point_exponential, backfit =TRUE)</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...</code></pre>
<pre><code>Warning in scale.EF(EF): Fitting stopped after the initialization function
failed to find a non-zero factor.</code></pre>
<pre><code>Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 6 factors (tolerance: 1.49e-04)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
  Difference between iterations is within 1.0e-03...
  Difference between iterations is within 1.0e-04...
Wrapping up...
Done.
Nullchecking 6 factors...
Done.</code></pre>
<p>Even with backfitting, some of the groups are put into the same
factors (it uses 6 factors for 10 groups); maybe not suprising because
the prior allows non-binary factors which can capture binary things…</p>
<pre class="r"><code>par(mfcol=c(2,3))
for(i in 1:6){
  plot(fit.bf$L_pm[,i])
}</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Examining the correlations of each factor with each group, we see the
6 inferred factors collectively pick out 10 groups, and each one picks
out 1-3 groups, as follows:</p>
<p>factor | group</p>
<p>1 | 3,7,8</p>
<p>2 | 5</p>
<p>3 | 6,10</p>
<p>4 | 1,2</p>
<p>5 | 9</p>
<p>6 | 4</p>
<pre class="r"><code>cor(L, fit.bf$L_pm)</code></pre>
<pre><code>             [,1]        [,2]         [,3]         [,4]        [,5]        [,6]
 [1,] -0.01234862 -0.10229226 -0.119340333  0.683062668  0.16349088 -0.09285511
 [2,]  0.09718390 -0.02964221  0.018222242  0.927376113  0.06491014 -0.11644674
 [3,]  0.69588688 -0.05316003 -0.166549542  0.006099336  0.06178274  0.05494743
 [4,]  0.12936739  0.11157517 -0.017243775 -0.127499408 -0.06270951  0.99734135
 [5,] -0.08711811  0.99931604 -0.191168934 -0.066146864  0.07510302  0.09816562
 [6,] -0.08431898 -0.14231333  0.855115116 -0.032825978  0.04329927  0.08336826
 [7,]  0.67526316  0.03908840  0.005572265  0.035581066 -0.06287441  0.17992081
 [8,]  0.53760926 -0.13504136 -0.026355268  0.127681105 -0.04054848 -0.03136272
 [9,] -0.03498443  0.08750939  0.045363376  0.109409666  0.99816824 -0.05168151
[10,] -0.09945685 -0.14236086  0.583478768 -0.032108437  0.04040719 -0.13044452</code></pre>
</div>
<div id="hierarchical-groups" class="section level2">
<h2>Hierarchical groups</h2>
<p>One thing I wondered: we found that it was helpful for trees to seek
a divergence factorization before the drift factorization, but now I’m
not entirely clear why this is. (I have some vague justifications,
including the fact that the resulting L vectors are closer to
orthogonal, but overall I’m not really sure why point-exponential does
not work so well). So here I try the point exponential with hierarchical
groups.</p>
<p>I generate some splits</p>
<pre class="r"><code>k = 6
set.seed(1)
l1 = rep(c(1,0),c(n/2,n/2))
l2 = rep(c(0,1),c(n/2,n/2))
l3 = rep(c(1,0,0,0),c(n/4,n/4,n/4,n/4))
l4 = rep(c(0,1,0,0),c(n/4,n/4,n/4,n/4))
l5 = rep(c(0,0,1,0),c(n/4,n/4,n/4,n/4))
l6 = rep(c(0,0,0,1),c(n/4,n/4,n/4,n/4))
L = cbind(l1,l2,l3,l4,l5,l6)
F = matrix(rnorm(k*p),nrow=p)
X = L %*% t(F) + rnorm(n*p,0,0.001)
fit = flash(X %*% t(X),ebnm_fn = ebnm_point_exponential, backfit=TRUE )</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...</code></pre>
<pre><code>Warning in scale.EF(EF): Fitting stopped after the initialization function
failed to find a non-zero factor.</code></pre>
<pre><code>Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 2 factors (tolerance: 1.49e-04)...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
Wrapping up...
Done.
Nullchecking 2 factors...
Done.</code></pre>
<p>We see that the fit stops after 2 factors. Here I plot the image of
the residuals. Green is negative numbers and red is positive.</p>
<pre class="r"><code>normalize = function(x){x/max(abs(x))}
image( normalize(residuals(fit)), levels=seq(-1,1,length=21),
      col= rgb(red = c(seq(0,1,length=10),1,rep(1,length=10)), 
               green = c(rep(1,length=10),1,seq(1,0,length=10)), 
               blue = c(seq(0,1,length=10),1,seq(1,0,length=10))) )</code></pre>
<pre><code>Warning in plot.window(...): &quot;levels&quot; is not a graphical parameter</code></pre>
<pre><code>Warning in plot.xy(xy, type, ...): &quot;levels&quot; is not a graphical parameter</code></pre>
<pre><code>Warning in axis(side = side, at = at, labels = labels, ...): &quot;levels&quot; is not a
graphical parameter

Warning in axis(side = side, at = at, labels = labels, ...): &quot;levels&quot; is not a
graphical parameter</code></pre>
<pre><code>Warning in box(...): &quot;levels&quot; is not a graphical parameter</code></pre>
<pre><code>Warning in title(...): &quot;levels&quot; is not a graphical parameter</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We see the problem: essentially flash removes “too much” of the top
branch, and can’t recover because of the non-negative constraint (and,
possibly, an over-estimated residual variance?). Indeed, it seems the
power method which is used to initialize does not find a non-negative
initial vector. This seems like a useful case study to try to fix.</p>
<p>I tried fixing the residual variance to see if it helped, but no
luck:</p>
<pre class="r"><code>fit = flash(X %*% t(X),ebnm_fn = ebnm_point_exponential, backfit=TRUE, var_type = NULL, S = 0.001 )</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...</code></pre>
<pre><code>Warning in scale.EF(EF): Fitting stopped after the initialization function
failed to find a non-zero factor.</code></pre>
<pre><code>Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 2 factors (tolerance: 1.49e-04)...
  Difference between iterations is within 1.0e+00...
Wrapping up...
Done.
Nullchecking 2 factors...
Done.</code></pre>
<p>Try point Laplace: it also doesn’t work here (doesn’t find sparse
solutions) perhaps because the small residual variance makes the fitting
sticky, and the initializations are not sparse enough?</p>
<pre class="r"><code>fit = flash(X %*% t(X),ebnm_fn = ebnm_point_laplace, backfit=TRUE)</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 4 factors (tolerance: 1.49e-04)...
  Difference between iterations is within 1.0e+04...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
Wrapping up...
Done.
Nullchecking 4 factors...
Done.</code></pre>
<pre class="r"><code>plot(fit$L_pm[,1])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(fit$L_pm[,2])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-13-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(fit$L_pm[,3])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-13-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(fit$L_pm[,4])</code></pre>
<p><img src="figure/flash_cov_overlapping_groups.Rmd/unnamed-chunk-13-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>I feel we might want to invest more in some simple ways to find
sparse initial solutions to initialize flashier with - eg using an L1
penalty and computing the full solution path? (We could implement L1
penalty both for the data matrix X and the covariance matrix XtX?) We
also need to understand the implications of over-removing the top branch
and maybe somehow to fix that. Maybe ideas related to nonnegative matrix
under-approximation, or to neighbor-joining (which has to work out how
to collapse two nodes into a single node) can help? Something to think
about: in a tree, does it help to find the sparsest factors first,
rather than the top branch, more similar to neighbor joining?</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] flashier_1.0.53 ebnm_1.1-2     

loaded via a namespace (and not attached):
 [1] httr_1.4.7           tidyr_1.3.0          sass_0.4.8          
 [4] viridisLite_0.4.2    jsonlite_1.8.8       splines_4.2.1       
 [7] RhpcBLASctl_0.23-42  gtools_3.9.5         bslib_0.6.1         
[10] RcppParallel_5.1.7   horseshoe_0.2.0      highr_0.10          
[13] mixsqp_0.3-54        deconvolveR_1.2-1    progress_1.2.3      
[16] yaml_2.3.8           ggrepel_0.9.4        fastTopics_0.6-175  
[19] pillar_1.9.0         lattice_0.22-5       quadprog_1.5-8      
[22] glue_1.6.2           digest_0.6.33        RColorBrewer_1.1-3  
[25] promises_1.2.1       colorspace_2.1-0     cowplot_1.1.3       
[28] htmltools_0.5.7      httpuv_1.6.13        Matrix_1.6-4        
[31] pkgconfig_2.0.3      invgamma_1.1         purrr_1.0.2         
[34] scales_1.3.0         whisker_0.4.1        later_1.3.2         
[37] Rtsne_0.17           git2r_0.33.0         tibble_3.2.1        
[40] generics_0.1.3       ggplot2_3.4.4        cachem_1.0.8        
[43] ashr_2.2-63          pbapply_1.7-2        lazyeval_0.2.2      
[46] cli_3.6.2            crayon_1.5.2         magrittr_2.0.3      
[49] evaluate_0.23        fs_1.6.3             fansi_1.0.6         
[52] truncnorm_1.0-9      prettyunits_1.2.0    tools_4.2.1         
[55] data.table_1.14.10   softImpute_1.4-1     hms_1.1.3           
[58] lifecycle_1.0.4      stringr_1.5.1        plotly_4.10.4       
[61] trust_0.1-8          munsell_0.5.0        irlba_2.3.5.1       
[64] compiler_4.2.1       jquerylib_0.1.4      rlang_1.1.2         
[67] grid_4.2.1           rstudioapi_0.15.0    htmlwidgets_1.6.4   
[70] rmarkdown_2.25       gtable_0.3.4         R6_2.5.1            
[73] knitr_1.45           dplyr_1.1.4          uwot_0.1.16         
[76] fastmap_1.1.1        utf8_1.2.4           workflowr_1.7.1     
[79] rprojroot_2.0.4      stringi_1.8.3        Polychrome_1.5.1    
[82] parallel_4.2.1       SQUAREM_2021.1       Rcpp_1.0.12         
[85] vctrs_0.6.5          scatterplot3d_0.3-44 tidyselect_1.2.0    
[88] xfun_0.41           </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
