<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Matthew Stephens" />

<meta name="date" content="2025-02-23" />

<title>nmu_em</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephens999/misc">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">nmu_em</h1>
<h4 class="author">Matthew Stephens</h4>
<h4 class="date">2025-02-23</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-26
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree19e41fe04813f5284a6bc98326603a48e1694bbbtargetblank19e41fea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephens999/misc/tree/19e41fe04813f5284a6bc98326603a48e1694bbb" target="_blank">19e41fe</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephens999misctree19e41fe04813f5284a6bc98326603a48e1694bbbtargetblank19e41fea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephens999/misc/tree/19e41fe04813f5284a6bc98326603a48e1694bbb" target="_blank">19e41fe</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/ALStruct_cache/
    Ignored:    data/.Rhistory
    Ignored:    data/methylation-data-for-matthew.rds
    Ignored:    data/pbmc/
    Ignored:    data/pbmc_purified.RData

Untracked files:
    Untracked:  .dropbox
    Untracked:  Icon
    Untracked:  analysis/GHstan.Rmd
    Untracked:  analysis/GTEX-cogaps.Rmd
    Untracked:  analysis/PACS.Rmd
    Untracked:  analysis/Rplot.png
    Untracked:  analysis/SPCAvRP.rmd
    Untracked:  analysis/abf_comparisons.Rmd
    Untracked:  analysis/admm_02.Rmd
    Untracked:  analysis/admm_03.Rmd
    Untracked:  analysis/bispca.Rmd
    Untracked:  analysis/cache/
    Untracked:  analysis/cholesky.Rmd
    Untracked:  analysis/compare-transformed-models.Rmd
    Untracked:  analysis/cormotif.Rmd
    Untracked:  analysis/cp_ash.Rmd
    Untracked:  analysis/eQTL.perm.rand.pdf
    Untracked:  analysis/eb_prepilot.Rmd
    Untracked:  analysis/eb_var.Rmd
    Untracked:  analysis/ebpmf1.Rmd
    Untracked:  analysis/ebpmf_sla_text.Rmd
    Untracked:  analysis/ebspca_sims.Rmd
    Untracked:  analysis/explore_psvd.Rmd
    Untracked:  analysis/fa_check_identify.Rmd
    Untracked:  analysis/fa_iterative.Rmd
    Untracked:  analysis/flash_cov_overlapping_groups_init.Rmd
    Untracked:  analysis/flash_test_tree.Rmd
    Untracked:  analysis/flashier_newgroups.Rmd
    Untracked:  analysis/flashier_nmf_triples.Rmd
    Untracked:  analysis/flashier_pbmc.Rmd
    Untracked:  analysis/flashier_snn_shifted_prior.Rmd
    Untracked:  analysis/greedy_ebpmf_exploration_00.Rmd
    Untracked:  analysis/ieQTL.perm.rand.pdf
    Untracked:  analysis/lasso_em_03.Rmd
    Untracked:  analysis/m6amash.Rmd
    Untracked:  analysis/mash_bhat_z.Rmd
    Untracked:  analysis/mash_ieqtl_permutations.Rmd
    Untracked:  analysis/methylation_example.Rmd
    Untracked:  analysis/mixsqp.Rmd
    Untracked:  analysis/mr.ash_lasso_init.Rmd
    Untracked:  analysis/mr.mash.test.Rmd
    Untracked:  analysis/mr_ash_modular.Rmd
    Untracked:  analysis/mr_ash_parameterization.Rmd
    Untracked:  analysis/mr_ash_ridge.Rmd
    Untracked:  analysis/mv_gaussian_message_passing.Rmd
    Untracked:  analysis/nejm.Rmd
    Untracked:  analysis/nmf_bg.Rmd
    Untracked:  analysis/nmu_split.Rmd
    Untracked:  analysis/nonneg_underapprox.Rmd
    Untracked:  analysis/normal_conditional_on_r2.Rmd
    Untracked:  analysis/normalize.Rmd
    Untracked:  analysis/pbmc.Rmd
    Untracked:  analysis/pca_binary_weighted.Rmd
    Untracked:  analysis/pca_l1.Rmd
    Untracked:  analysis/poisson_nmf_approx.Rmd
    Untracked:  analysis/poisson_shrink.Rmd
    Untracked:  analysis/poisson_transform.Rmd
    Untracked:  analysis/qrnotes.txt
    Untracked:  analysis/ridge_iterative_02.Rmd
    Untracked:  analysis/ridge_iterative_splitting.Rmd
    Untracked:  analysis/samps/
    Untracked:  analysis/sc_bimodal.Rmd
    Untracked:  analysis/shrinkage_comparisons_changepoints.Rmd
    Untracked:  analysis/susie_cov.Rmd
    Untracked:  analysis/susie_en.Rmd
    Untracked:  analysis/susie_z_investigate.Rmd
    Untracked:  analysis/svd-timing.Rmd
    Untracked:  analysis/temp.RDS
    Untracked:  analysis/temp.Rmd
    Untracked:  analysis/test-figure/
    Untracked:  analysis/test.Rmd
    Untracked:  analysis/test.Rpres
    Untracked:  analysis/test.md
    Untracked:  analysis/test_qr.R
    Untracked:  analysis/test_sparse.Rmd
    Untracked:  analysis/tree_dist_top_eigenvector.Rmd
    Untracked:  analysis/z.txt
    Untracked:  code/multivariate_testfuncs.R
    Untracked:  code/rqb.hacked.R
    Untracked:  data/4matthew/
    Untracked:  data/4matthew2/
    Untracked:  data/E-MTAB-2805.processed.1/
    Untracked:  data/ENSG00000156738.Sim_Y2.RDS
    Untracked:  data/GDS5363_full.soft.gz
    Untracked:  data/GSE41265_allGenesTPM.txt
    Untracked:  data/Muscle_Skeletal.ACTN3.pm1Mb.RDS
    Untracked:  data/P.rds
    Untracked:  data/Thyroid.FMO2.pm1Mb.RDS
    Untracked:  data/bmass.HaemgenRBC2016.MAF01.Vs2.MergedDataSources.200kRanSubset.ChrBPMAFMarkerZScores.vs1.txt.gz
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.NewSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.PreviousSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/eb_prepilot/
    Untracked:  data/finemap_data/fmo2.sim/b.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/data
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.ld
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.z
    Untracked:  data/finemap_data/fmo2.sim/pos.txt
    Untracked:  data/logm.csv
    Untracked:  data/m.cd.RDS
    Untracked:  data/m.cdu.old.RDS
    Untracked:  data/m.new.cd.RDS
    Untracked:  data/m.old.cd.RDS
    Untracked:  data/mainbib.bib.old
    Untracked:  data/mat.csv
    Untracked:  data/mat.txt
    Untracked:  data/mat_new.csv
    Untracked:  data/matrix_lik.rds
    Untracked:  data/paintor_data/
    Untracked:  data/running_data_chris.csv
    Untracked:  data/running_data_matthew.csv
    Untracked:  data/temp.txt
    Untracked:  data/y.txt
    Untracked:  data/y_f.txt
    Untracked:  data/zscore_jointLCLs_m6AQTLs_susie_eQTLpruned.rds
    Untracked:  data/zscore_jointLCLs_random.rds
    Untracked:  explore_udi.R
    Untracked:  output/fit.k10.rds
    Untracked:  output/fit.nn.pbmc.purified.rds
    Untracked:  output/fit.nn.rds
    Untracked:  output/fit.nn.s.001.rds
    Untracked:  output/fit.nn.s.01.rds
    Untracked:  output/fit.nn.s.1.rds
    Untracked:  output/fit.nn.s.10.rds
    Untracked:  output/fit.snn.s.001.rds
    Untracked:  output/fit.snn.s.01.nninit.rds
    Untracked:  output/fit.snn.s.01.rds
    Untracked:  output/fit.varbvs.RDS
    Untracked:  output/fit2.nn.pbmc.purified.rds
    Untracked:  output/glmnet.fit.RDS
    Untracked:  output/snn07.txt
    Untracked:  output/snn34.txt
    Untracked:  output/test.bv.txt
    Untracked:  output/test.gamma.txt
    Untracked:  output/test.hyp.txt
    Untracked:  output/test.log.txt
    Untracked:  output/test.param.txt
    Untracked:  output/test2.bv.txt
    Untracked:  output/test2.gamma.txt
    Untracked:  output/test2.hyp.txt
    Untracked:  output/test2.log.txt
    Untracked:  output/test2.param.txt
    Untracked:  output/test3.bv.txt
    Untracked:  output/test3.gamma.txt
    Untracked:  output/test3.hyp.txt
    Untracked:  output/test3.log.txt
    Untracked:  output/test3.param.txt
    Untracked:  output/test4.bv.txt
    Untracked:  output/test4.gamma.txt
    Untracked:  output/test4.hyp.txt
    Untracked:  output/test4.log.txt
    Untracked:  output/test4.param.txt
    Untracked:  output/test5.bv.txt
    Untracked:  output/test5.gamma.txt
    Untracked:  output/test5.hyp.txt
    Untracked:  output/test5.log.txt
    Untracked:  output/test5.param.txt

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/flashier_log1p.Rmd
    Modified:   analysis/flashier_sla_text.Rmd
    Modified:   analysis/logistic_z_scores.Rmd
    Modified:   analysis/mr_ash_pen.Rmd
    Modified:   analysis/susie_flash.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/nmu_em.Rmd</code>) and HTML
(<code>docs/nmu_em.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephens999/misc/blob/19e41fe04813f5284a6bc98326603a48e1694bbb/analysis/nmu_em.Rmd" target="_blank">19e41fe</a>
</td>
<td>
Matthew Stephens
</td>
<td>
2025-02-26
</td>
<td>
wflow_publish("nmu_em.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>My goal is to fit a version of the non-negative matrix
underapproximation using an EM algorithm.</p>
<p>The model is: <span class="math display">\[A \sim uv&#39; + b +
e\]</span> where <span class="math inline">\(b \sim
Exp(\lambda)\)</span> and <span class="math inline">\(e \sim
N(0,sigma^2)\)</span>. Or in other words, <span class="math inline">\(A
\sim N(uv&#39; + b,\sigma^2)\)</span>. If <span
class="math inline">\(\sigma^2=0\)</span> then the mle for u,v should be
a feasible solution to the underapproximation problem. Introducing <span
class="math inline">\(\sigma\)</span> allows us to implement an EM
algorithm.</p>
<p>More generally, the intuition is that if <span
class="math inline">\(sigma^2\)</span> is very small (compared with
<span class="math inline">\(1/\lambda\)</span>) then this will
approximately solve (a version of) the non-negative matrix
underapproximation problem. If <span
class="math inline">\(1/\lambda\)</span> is very small compared with
<span class="math inline">\(\sigma\)</span> then it will be closer to
regular NMF.</p>
<p>NOTE: we could even use this idea within flashier to put priors on u
and v…</p>
<p>The ELBO is <span class="math display">\[F(u,v,q)=
E_q((-1/2\sigma^2)||A-uv&#39;-b||_2^2) + D_{KL}(q,g)\]</span> where
<span class="math inline">\(g\)</span> is the prior on <span
class="math inline">\(b\)</span>, <span
class="math display">\[g(b)=\lambda \exp(-\lambda b).\]</span> Here
<span class="math inline">\(q\)</span> plays the role of the
(approximate) posterior distribution on <span
class="math inline">\(b\)</span>.</p>
<p>Given <span class="math inline">\(q\)</span>, this is minimized for
<span class="math inline">\(u,v\)</span> by solving <span
class="math display">\[\min_{u,v} ||A-\bar{b} - uv&#39;||_2^2\]</span>
For now we will do this approximately, by thresholding the rank 1 svd of
<span class="math inline">\(A-\bar{b}\)</span>. NOTE: I do this
currently by thresholding udv’; I could instead try thresholding u and v
(which is more in keeping with requiring them to be non-negative) and
should probably try that later.</p>
<p>Given <span class="math inline">\(u,v\)</span> this is minimized by
for each <span class="math inline">\(b_{ij}\)</span> by solving <span
class="math display">\[q(b) \propto g(b)
\exp((-1/2\sigma^2)(x_{ij}-b)^2) \propto
\exp((-1/2\sigma^2)[b^2-2(x_{ij}-\lambda \sigma^2)b])\]</span> where
<span class="math inline">\(x_{ij} = A_ij - u_i v_j\)</span>. This is a
truncated normal distribution, <span class="math display">\[  q(b_{ij})
= N_+(x_{ij}-\lambda \sigma^2, \sigma^2)\]</span>.</p>
<p>Example: Suppose <span class="math inline">\(\sigma\)</span> is very
small. Suppose <span class="math inline">\(x=-0.5\)</span>, and we have
<span class="math inline">\(x \sim N(b,\sigma^2)\)</span> and <span
class="math inline">\(b \sim Exp(1)\)</span>.</p>
<pre class="r"><code>  x=-0.5
  b = seq(0,1,length=100)
  logprior = dexp(b,log=TRUE)
  loglik = dnorm(-0.5,b,sd=1,log=TRUE)
  plot(exp(logprior+loglik))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  loglik = dnorm(-0.5,b, sd=0.1, log=TRUE)
  plot(exp(logprior+loglik))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-1-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  sigma=1  
  x = seq(-0.5,0.5,length=20)
  plot(x,truncnorm::etruncnorm(0,Inf,x-sigma^2,sigma))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-1-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  sigma=0.01  
  x = seq(-0.5,0.5,length=20)
  plot(x,truncnorm::etruncnorm(0,Inf,x-sigma^2,sigma))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-1-4.png" width="672" style="display: block; margin: auto;" /></p>
<div id="approach-1-thresholding-m" class="section level3">
<h3>Approach 1: thresholding M</h3>
<p>Let’s try. First I simulate some data for testing -</p>
<pre class="r"><code>set.seed(1)
n = 10
maxiter = 1000
x = cbind(c(rep(1,n),rep(0,n)), c(rep(0,n),rep(1,n)))
E = matrix(0.1*rexp(2*n*2*n),nrow=2*n)
E = E+t(E) #symmetric errors
A = x %*% t(x) + E</code></pre>
<p>This is a first try with lambda=sigma=1. This didn’t do exactly what
I wanted (which is an under-approximation), but maybe not surprising as
lambda and sigma are similar.</p>
<pre class="r"><code>lambda = 1
sigma = 1
b = matrix(0,nrow=nrow(A),ncol=ncol(A))

for(i in 1:maxiter){
  A.svd = svd(A-b)
  M = A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1])
  M = ifelse(M&lt;0,0,M)
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-M-lambda*sigma^2,sd=sigma),nrow=2*n)
}
image(M)</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now I try sigma small, to try to put the errors in the exponential
part and force an underapproximation. This work better, but it is not
strictly an underapproximation.</p>
<pre class="r"><code>lambda = 1
sigma = .1
b = matrix(0,nrow=nrow(A),ncol=ncol(A))

for(i in 1:maxiter){
  A.svd = svd(A-b)
  M = A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1])
  M = ifelse(M&lt;0,0,M)
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-M-lambda*sigma^2,sd=sigma),nrow=2*n)
}
image(M)</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A-b)</code></pre>
<pre><code>[1] -0.07059372</code></pre>
<pre class="r"><code>min(A-A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1]))</code></pre>
<pre><code>[1] 0.02239963</code></pre>
<pre class="r"><code>lambda = 1
sigma = .01
b = matrix(0,nrow=nrow(A),ncol=ncol(A))

for(i in 1:maxiter){
  A.svd = svd(A-b)
  M = A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1])
  M = ifelse(M&lt;0,0,M)
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-M-lambda*sigma^2,sd=sigma),nrow=2*n)
}
image(M)</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A-b)</code></pre>
<pre><code>[1] -0.003307836</code></pre>
<pre class="r"><code>min(A-A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1]))</code></pre>
<pre><code>[1] 0.008896136</code></pre>
<p>So this is still not an underapproximation (and also u and v are not
non-negative).</p>
<pre class="r"><code>A.svd$u[,1]</code></pre>
<pre><code> [1] -3.532064e-01 -3.408483e-01 -2.994702e-01 -3.289122e-01 -3.005021e-01
 [6] -2.694988e-01 -2.925585e-01 -2.878785e-01 -3.508508e-01 -3.179764e-01
[11] -4.226460e-02 -8.810541e-03  9.533596e-05 -4.853362e-03  5.891319e-05
[16] -4.262211e-02 -1.151787e-02 -2.184533e-02 -3.587380e-02 -5.177057e-04</code></pre>
<p>Before moving on to thresholding u and v, I note that the result can
depend on the scale of lambda, sigma. We might want to frame the problem
a bit differently to avoid that… eg by <span class="math inline">\(A =
\sigma(uv&#39;+b+e)\)</span> where <span class="math inline">\(e \sim
N(0,1)\)</span> and <span class="math inline">\(b \sim
Exp(\lambda)\)</span>. This would make the approach scale invariant,
which seems preferable. Interestingly this one is an underapproximation.
It might be worth investigating a bit more when both lambda and sigma
are big…</p>
<pre class="r"><code>lambda = 100
sigma = 100
b = matrix(0,nrow=nrow(A),ncol=ncol(A))

for(i in 1:maxiter){
  A.svd = svd(A-b)
  M = A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1])
  M = ifelse(M&lt;0,0,M)
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-M-lambda*sigma^2,sd=sigma),nrow=2*n)
}
image(M)</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A-b)</code></pre>
<pre><code>[1] -0.003282601</code></pre>
<pre class="r"><code>min(A-A.svd$d[1]* A.svd$u[,1] %*% t(A.svd$v[,1]))</code></pre>
<pre><code>[1] -0.6801494</code></pre>
</div>
<div id="approach-2-thresholding-uv" class="section level3">
<h3>Approach 2: thresholding u,v</h3>
<p>First I’m going to solve nmf by power method. The initialization here
shoudl probably be checked more carefully.</p>
<pre class="r"><code># solves the rank 1 symmetric nmf problem by performing niter iterations of the (thresholded) power method
# init should be a list with named element u that is used for initialization
# if not supplied uses svd followed by truncation to initialize u.
symnmf_r1 = function(A,init = NULL, niter=1){
  
  if(is.null(init)){ #initialize by thresholding first singular vector
    A.svd = svd(A)
    u = A.svd$u[,1]  #note that thresholding u and -u give different results, so i try both ways and choose whichever gives the larger value of u&#39;Au
    u0a = ifelse(u&lt;0,0,u)
    if(!all(u0a==0))
      u0a = u0a/sqrt(sum(u0a^2))
     
    u0b = ifelse(u&gt;0,0,-u) #corresponds to using -u instead of u
    if(!all(u0b==0))
      u0b = u0b/sqrt(sum(u0b^2))
    
    fa = t(u0a) %*% A %*% u0a
    fb = t(u0b) %*% A %*% u0b
    
    if(fa&gt;fb)
      u = u0a
    else
      u = u0b
  } 
  else {
    u = init$u
  }
  
  for(i in 1:niter){
    u = A %*% u
    u = ifelse(u&lt;0,0,u)
    u = u/sqrt(sum(u^2))
  }
  
  u = u/sqrt(sum(u^2))
  
  v = (A %*% u)
  v = ifelse(v&lt;0,0,v)
  d = sqrt(sum(v^2))
  v = v/sqrt(sum(v^2))
  return(list(u=u,d=d,v=v))
}

  
B = A-min(A)
temp = symnmf_r1(B)
temp2 = symnmf_r1(A)</code></pre>
<pre class="r"><code>lambda = 1
sigma = 1
b = matrix(0,nrow=nrow(A),ncol=ncol(A))
temp = symnmf_r1(A-b)

for(i in 1:maxiter){
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-temp$d*temp$u %*% t(temp$v)-lambda*sigma^2,sd=sigma),nrow=2*n)
  temp = symnmf_r1(A-b,temp)
}
image(temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<pre><code>[1] 0.008737816</code></pre>
<pre class="r"><code>hist(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-9-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lambda = 1
sigma = .1
b = matrix(0,nrow=nrow(A),ncol=ncol(A))
temp = symnmf_r1(A-b)

for(i in 1:maxiter){
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-temp$d*temp$u %*% t(temp$v)-lambda*sigma^2,sd=sigma),nrow=2*n)
  temp = symnmf_r1(A-b,temp)
}
image(temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<pre><code>[1] 0.008737816</code></pre>
<pre class="r"><code>hist(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lambda = 1
sigma = .01
b = matrix(0,nrow=nrow(A),ncol=ncol(A))
temp = symnmf_r1(A-b)

for(i in 1:maxiter){
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-temp$d*temp$u %*% t(temp$v)-lambda*sigma^2,sd=sigma),nrow=2*n)
  temp = symnmf_r1(A-b,temp)
}
image(temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<pre><code>[1] 0.008737816</code></pre>
<pre class="r"><code>hist(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-11-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(A-temp$d * temp$u %*% t(temp$v), b)</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-11-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lambda = 100
sigma = 100
b = matrix(0,nrow=nrow(A),ncol=ncol(A))
temp = symnmf_r1(A-b)

for(i in 1:maxiter){
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-temp$d*temp$u %*% t(temp$v)-lambda*sigma^2,sd=sigma),nrow=2*n)
  temp = symnmf_r1(A-b,temp)
}
image(temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<pre><code>[1] -0.6806203</code></pre>
<pre class="r"><code>hist(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-12-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>All these look pretty good, except the last one (which is a bit weird
anyway).</p>
<p>Note: i did try applying this, accidentally, to a matrix where some A
were negative, so there is no underapproximation solution. It still did
something sensible!</p>
</div>
</div>
<div id="estimating-lambda-and-sigma" class="section level2">
<h2>Estimating lambda and sigma</h2>
<p>Note that <span class="math display">\[E(A-uv&#39;) =
1/\lambda\]</span> and <span class="math display">\[E(A-uv&#39;)^2) =
1/\lambda^2 + \sigma^2\]</span>.</p>
<p>So a method of moments suggests estimating <span
class="math inline">\(\lambda = 1/mean(A-uv&#39;)\)</span> and <span
class="math inline">\(\sigma^2 = mean((A-uv&#39;)^2) - mean(A-uv&#39;)^2
= var(A-uv&#39;)\)</span>.</p>
<pre class="r"><code>for(j in 1:10){
  lambda = 1/mean(A-temp$d*temp$u %*% t(temp$v))
  sigma = sd(A-temp$d*temp$u %*% t(temp$v))
  b = matrix(0,nrow=nrow(A),ncol=ncol(A))

  for(i in 1:maxiter){
    temp = symnmf_r1(A-b)
    b = matrix(truncnorm::etruncnorm(a=0,mean= A-temp$d*temp$u %*% t(temp$v)-lambda*sigma^2,sd=sigma),nrow=2*n)
  }
}

image(temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<pre><code>[1] -0.6700389</code></pre>
<pre class="r"><code>hist(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-13-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lambda</code></pre>
<pre><code>[1] 47.00306</code></pre>
<pre class="r"><code>sigma</code></pre>
<pre><code>[1] 0.525397</code></pre>
</div>
<div id="reducing-sigma" class="section level2">
<h2>Reducing sigma</h2>
<p>I’d like to try to get an actual under-approximation by reducing
sigma. Here I will iteratively reduce sigma. (NOTE: i did this when I
was not getting an underapproximation due to A having negative values..
might not be necessary…)</p>
<pre class="r"><code>lambda = 1
sigma = 1
b = matrix(0,nrow=nrow(A),ncol=ncol(A))

for(i in 1:100){
  temp = symnmf_r1(A-b)
  b = matrix(truncnorm::etruncnorm(a=0,mean= A-temp$d*temp$u %*% t(temp$v)-lambda*sigma^2,sd=sigma),nrow=2*n)
  sigma = sigma/1.1
}

image(temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>min(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<pre><code>[1] 0.008677423</code></pre>
<pre class="r"><code>hist(A- temp$d*temp$u %*% t(temp$v))</code></pre>
<p><img src="figure/nmu_em.Rmd/unnamed-chunk-14-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="scaling-solution" class="section level2">
<h2>scaling solution</h2>
<p>Now I am going to try <span class="math display">\[A = \sigma(uv&#39;
+ b + e)\]</span> where <span class="math inline">\(u,v\)</span>
non-negative, <span class="math inline">\(e \sim N(0,1)\)</span> and
<span class="math inline">\(b\sim Exp(lambda)\)</span>.</p>
<p>The ELBO is <span class="math display">\[F(u,v,q)=
E_q((-1/2\sigma^2)||A-\sigma uv&#39;-\sigma b||_2^2) +
D_{KL}(q,g)\]</span> where <span class="math inline">\(g\)</span> is the
prior on <span class="math inline">\(b\)</span>, <span
class="math display">\[g(b)=\lambda \exp(-\lambda b).\]</span> Here
<span class="math inline">\(q\)</span> plays the role of the
(approximate) posterior distribution on <span
class="math inline">\(b\)</span>.</p>
<p>Given <span class="math inline">\(q\)</span>, this is minimized for
<span class="math inline">\(u,v\)</span> by solving <span
class="math display">\[\min_{u,v} ||A/\sigma -\bar{b} -
uv&#39;||_2^2\]</span></p>
<p>Given <span class="math inline">\(u,v\)</span> this is minimized by
for each <span class="math inline">\(b_{ij}\)</span> by solving <span
class="math display">\[q(b) \propto g(b) \exp((-1/2)(x_{ij} - b)^2)
\propto \exp((-1/2)[b^2-2(x_{ij}-\lambda)b])\]</span> where <span
class="math inline">\(x_{ij} = A_ij/\sigma - u_i v_j\)</span>. This is a
truncated normal distribution, <span class="math display">\[q(b_{ij}) =
N_+(x_{ij}-\lambda, 1)\]</span>.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5       cli_3.6.3         knitr_1.49        rlang_1.1.5      
 [5] xfun_0.50         stringi_1.8.4     promises_1.3.2    jsonlite_1.8.9   
 [9] workflowr_1.7.1   glue_1.8.0        rprojroot_2.0.4   git2r_0.35.0     
[13] htmltools_0.5.8.1 httpuv_1.6.15     sass_0.4.9        rmarkdown_2.29   
[17] evaluate_1.0.3    jquerylib_0.1.4   tibble_3.2.1      fastmap_1.2.0    
[21] yaml_2.3.10       lifecycle_1.0.4   whisker_0.4.1     stringr_1.5.1    
[25] compiler_4.4.2    fs_1.6.5          Rcpp_1.0.14       pkgconfig_2.0.3  
[29] rstudioapi_0.17.1 later_1.4.1       truncnorm_1.0-9   digest_0.6.37    
[33] R6_2.5.1          pillar_1.10.1     magrittr_2.0.3    bslib_0.9.0      
[37] tools_4.4.2       cachem_1.1.0     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
